{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 562,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 562,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "import os\n",
    "from getpass import getpass\n",
    "\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "from langchain.embeddings import HuggingFaceBgeEmbeddings\n",
    "from langchain.chains import RetrievalQA\n",
    "from langchain.retrievers import ContextualCompressionRetriever\n",
    "from langchain.retrievers.document_compressors import CohereRerank\n",
    "from langchain_community.document_loaders import UnstructuredMarkdownLoader\n",
    "from langchain_pinecone import PineconeVectorStore\n",
    "from langchain_mistralai import ChatMistralAI\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain.output_parsers import PydanticOutputParser\n",
    "from langchain_core.prompts import PromptTemplate\n",
    "from langchain_core.pydantic_v1 import BaseModel, Field, validator\n",
    "from dotenv import load_dotenv\n",
    "from langchain.vectorstores import FAISS\n",
    "from typing import List,Optional\n",
    "import pandas as pd\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 563,
   "metadata": {},
   "outputs": [],
   "source": [
    "OPENAI_TOKEN = os.getenv('OPENAI_TOKEN')\n",
    "COHERE_TOKEN = os.getenv('COHERE_TOKEN')\n",
    "MISTRAL_TOKEN = os.getenv('MISTRAL_TOKEN')\n",
    "os.environ['PINECONE_API_KEY'] = os.getenv('PINECONE')\n",
    "file_path = 'processed_data/s41418-019-0422-6.md'\n",
    "fule_name = 's41418-019-0422-6'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "metadata": {},
   "outputs": [],
   "source": [
    "loader = UnstructuredMarkdownLoader(file_path)\n",
    "loaded_documents = loader.load()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 565,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 565,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_splitter = RecursiveCharacterTextSplitter(chunk_size=8000, chunk_overlap=1000)\n",
    "docs = text_splitter.split_documents(loaded_documents)\n",
    "len(docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "metadata": {},
   "outputs": [],
   "source": [
    "#model_name = 'intfloat/multilingual-e5-large'\n",
    "#model_name = 'BAAI/bge-small-en-v1.5'\n",
    "model_name = 'BAAI/bge-m3'\n",
    "#model_name = 'dunzhang/stella_en_1.5B_v5'\n",
    "\n",
    "#model_name = 'sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_name = model_name\n",
    "encode_kwargs = {'normalize_embeddings': True} # set True to compute cosine similarity\n",
    "\n",
    "embeddings = HuggingFaceBgeEmbeddings(\n",
    "    model_name=model_name,\n",
    "    model_kwargs={'device': 'mps'},\n",
    "    encode_kwargs=encode_kwargs\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 548,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"What groups of animals are used in this study?\n",
    "         \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 549,
   "metadata": {},
   "outputs": [],
   "source": [
    "vectorstore_from_docs = FAISS.from_documents(docs,\n",
    "        embedding=embeddings\n",
    "    )\n",
    "retriever = vectorstore_from_docs.as_retriever(search_kwargs={\"k\": 10})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 550,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pretty_print_docs(docs):\n",
    "    print(\n",
    "        f\"\\n{'-' * 100}\\n\".join(\n",
    "            [f\"Document {i+1}:\\n\\n\" + d.page_content for i, d in enumerate(docs)]\n",
    "        )\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 551,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Document 1:\n",
      "\n",
      "Statistical analysis\n",
      "\n",
      "All statistical analyses were performed using GraphPad Prism software version 6 (GraphPad software Inc). No statistical methods were used to predetermine sample sizes, but our sample sizes are similar to or greater than those reported in previous publications [35–37]. Data represent the mean and standard error of the mean (SEM). Unpaired two-tailed Student’s t test was used for the comparison of two means. ANOVA followed by Bonferroni’s or Tukey’s post hoc test were used for the multiple group analysis. Log-rank test was used for disease onset and survival analysis. The significance level for the two-sided analyses was set at P < 0.05.\n",
      "\n",
      "Results\n",
      "\n",
      "CNS-specific LanCL1 transgene expression improves the survival and motor function of ALS mice\n",
      "\n",
      "As a first step towards the understanding of the role of LanCL1 in ALS, we characterized the temporal expression patterns of LanCL1 in SOD1G93A mice. Since a decline in body weight is highly correlated with denervation-induced muscle atrophy, weight loss is well established as a simple and objective indicator of the disease course [29]. Thus, we analyzed the disease course based on body weight alterations (Fig. 1a). To evaluate LanCL1 expression at different stages of the disease, both mRNA and protein of lumbar spinal cords from SOD1G93A (G93A) mice and WT littermates were analyzed at the presymptomatic stage (P30 and P75), Early-sym stage (P120) and Late-sym stage (P150). Results showed that both LanCL1 mRNA and protein levels were increased at the late presymptomatic stage (P75) as compared with WT mice (Fig. 1b–d). However, with disease development, the mRNA and protein levels of LanCL1 gradually decreased in SOD1G93A spinal cord compared with WT (Fig. 1b–d).\n",
      "\n",
      "To examine whether LanCL1 confers neuroprotective effect in ALS mice (SOD1G93A), we crossed CNS-specific LanCL1 transgenic mice (LanCL1K/K; Nestin Cre+/−) with ALS mice. Four experimental groups were contrasted as follows: (1) SOD1G93A mice with one or two copies of LanCL1 transgene driven by Nestin-Cre (henceforth referred to as G93A; LanCL1 cKI); (2) SOD1G93A mice without LanCL1 transgene (henceforth referred to as G93A); (3) CNS-specific LanCL1 overexpression without SOD1 G93A (henceforth referred to as LanCL1 cKI); (4) neither LanCL1 transgene nor mutant SOD1 is expressed (henceforth referred to as WT; Fig. 2a).\n",
      "\n",
      "We monitored the survival of G93A and G93A; LanCL1 cKI mice. In males, CNS-specific LanCL1 transgene resulted in a 22 days extension in median survival (141 days for G93A and 163 days for G93A; LanCL1 cKI, respectively; Fig. 2b). Similarly, among females, median survival was increased from 146 days for G93A mice to 168 days for G93A; LanCL1 cKI mice (p < 0.01) (Fig. 2c). Using peak body weight to determine disease onset [38] (Fig. 1a), we found that LanCL1 transgene delayed disease onset by ~13% (p < 0.05) among males, from 105 days for G93A mice to 119 days for G93A; LanCL1 cKI mice (Fig. 2d), and by ~16% (p < 0.05) for females, from 108 days for G93A mice to 125.5 days for G93A; LanCL1 cKI animals (Fig. 2e). In addition, LanCL1 transgene extended the progression of the disease by ~43% (p < 0.01) and ~24% (p < 0.01) for males and females, respectively (Fig. 2f, g).\n",
      "\n",
      "Motor impairment is a prominent feature of ALS mouse model. To examine whether motor deficits in ALS mice was ameliorated by LanCL1 transgene, we analyzed the motor...# H. Tan et al.\n",
      "\n",
      "SOD\n",
      "\n",
      "LanCL\n",
      "\n",
      "Nestin\n",
      "\n",
      "Experimental animals\n",
      "\n",
      "Males Males Males 1C0: 8 2 20 L 89 105 0.01 6934 (n=15) G934 (n=17) 6934 LanCL cKi (n=15) 100 200 100 1C0 150 6934 6934 LanCL cKI\n",
      "\n",
      "Females Females Females 100 2 20 3 0.01 146 38 0.05 125.5 L 6934 (n=21) G934 (n=18) cKi (n=16) G934 LanCL cKi (n=16) 100 200 100 150 G934 6934 anCli cki\n",
      "\n",
      "Males Females 350 350 300 300 250 250 200 200 150 150 100 100 G93A (n=13) C93A (n=12) G934 LanCLI cKl (n=13) 6934 LanCLi cKI (n=11) 11 12 13 14 15 16 17 18 12 13 14 15 16 17 18 Age (weeks) Age (weeks)\n",
      "\n",
      "Function by Rotarod assay. G93A; LanCL1 cKI mice exhibited better performance on Rotarod tests at all time points compared with G93A mice, beginning as early as postnatal day 63 (P63) for females and P70 for males (Fig. 2h, i). Age showed that one or two copies of LanCL1 transgene did not.# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "Fig. 2\n",
      "\n",
      "CNS-specific expression of LanCL1 transgene significantly prolongs the lifespan, delays disease onset, extends disease progression, and improves motor function of ALS mice.\n",
      "\n",
      "a Breeding strategy for the generation of G93A; LanCL1 cKI triple-transgenic mice and control littermates. Kaplan–Meier curves of the survival showing that LanCL1 overexpression increases median survival\n",
      "\n",
      "G93A males (n = 15 mice) G93A; LanCL1 cKI males (n = 15 mice) Median Survival 141 days 163 days\n",
      "\n",
      "b from 146 days for G93A females (n = 21 mice) to 168 days for G93A; LanCL1 cKI females (n = 16 mice). **p < 0.01, by log-rank test. Kaplan–Meier curves of disease onset showing that LanCL1 overexpression delays disease onset\n",
      "\n",
      "G93A males (n = 17 mice) G93A; LanCL1 cKI males (n = 15 mice) Disease Onset 105 days 119 days\n",
      "\n",
      "d from 108 days for G93A females (n = 18 mice) to 125.5 days for G93A; LanCL1 cKI females (n = 16 mice). *p < 0.05, by log-rank test. Mean duration of disease (days from onset to endpoint) showing that LanCL1 overexpression extends disease progression\n",
      "\n",
      "G93A males (n = 16 mice) G93A; LanCL1 cKI males (n = 15 mice) Duration of Disease 30.4 days 43 days\n",
      "\n",
      "f from 37.5 days for G93A females (n = 17 mice) to 46.3 days for G93A; LanCL1 cKI females (n = 18 mice). p < 0.01, by two-tailed unpaired Student’s t test. Rotarod tests showing that LanCL1 overexpression improves motor function of male (h) and female (i) G93A mice. Data represent mean ± SEM, n = 11–13 mice per sex per genotype, p < 0.05, p < 0.001, by Mann–Whitney U-tests\n",
      "\n",
      "Fig. 3\n",
      "\n",
      "LanCL1 overexpression rescues MN loss in the spinal cord of ALS mice. Representative images (a) and quantifications (b) of Nissl-stained MNs (indicated by white arrows) in matching lumbar ventral horn (VH) at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test. Representative make a significant difference. Western blotting showed that a single copy of LanCL1 transgene substantially increased LanCL1 protein level in the spinal cord relative to WT tissue, and two copies of LanCL1 did not produce additional increase in LanCL1 protein (Supplementary Fig. 1a, b).\n",
      "\n",
      "LanCL1 overexpression delays MN loss in ALS mice\n",
      "\n",
      "The improved performance in these behavioral tests suggests that the MN degeneration in the SOD1 mutant mice might be delayed by LanCL1 transgene. To examine this notion, we compared MN survival in G93A; LanCL1 cKI and G93A mice, relative to that in WT littermates by performing Nissl staining in the spinal VH of these comparable mice at Early-sym stage (P110) and Late-sym stage (P135),\n",
      "\n",
      "G93A LanCL1 cKI WT P110 P135\n",
      "\n",
      "images (c) and quantifications (d) of SMI-32-stained alpha MNs (indicated by white arrows) in matching lumbar ventral horn at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test.# H. Tan et al.\n",
      "\n",
      "WT LanCLI ckl 693A 693A LanCL1 cKi WT P110 LanCLI cki 6934 G934 LanCLi cki n-3 Vh 5 P135 20 Hm [ 8] 3 2 P11O P135\n",
      "\n",
      "Fig. 4 LanCL1 overexpression decreases neuroinflammation in the spinal cord of ALS mice.\n",
      "\n",
      "a Immunostaining of GFAP (green) and DAPI (blue) in matching lumbar ventral horn (VH) at P110 and P135.\n",
      "\n",
      "b Quantification of GFAP immunoreactivity in the lumbar ventral horn of indicated mice at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test respectively.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 2:\n",
      "\n",
      "Here, we report that LanCL1 protects MNs against degeneration and reduces the severity of disease manifestations in the mouse model of ALS. By crossing LanCL1 conditional transgenic mice with the SOD1G93A mouse model, we found that CNS-specific expression of LanCL1 transgene significantly prolongs lifespan, delays disease onset, decelerates symptomatic progression, and improves motor performance in SOD1G93A mice. Reciprocally, CNS-specific deletion of LanCL1 causes neurodegenerative phenotypes, including MN loss, neuroinflammation, and oxidative damage in spinal cord. We further show that loss of LanCL1 leads to a decrease in AKT phosphorylation, whereas LanCL1 transgene restores AKT phosphorylation and mitigates oxidative stress in ALS mice. Findings suggest that LanCL1 modulates MN survival by scavenging ROS and enhancing AKT activity.\n",
      "\n",
      "Material and methods\n",
      "\n",
      "Animals\n",
      "\n",
      "The generation and characterization of the LanCL1 conditional knockin (termed LanCL1 cKI) mice and LanCL1flox/flox mice have been previously described [26]. SOD1G93A [B6SJL-TgN(SOD1-G93A)1Gur] and Nestin-Cre (C57BL/6) mice were originally derived from the Jackson Laboratory (Bar Harbor, Maine, USA) and maintained on C57BL/6 and 129s4 mixed background. Male mice heterozygous for the human SOD1G93A Cre+/− LanCL1K/K; Nestin transgene were crossbred with female mice, resulting in triple-transgenic mice and control littermates. Littermates were always used as controls. The breeding strategy and the resulting genotypes are reported in Fig. 2a. Mice carrying the floxed allele of LanCL1 were crossed to Nestin-Cre transgenic mice [28] to generate LanCL1 conditional knockout (termed LanCL1 cKO) mice with LanCL1 deficiency in central nervous system (CNS). LanCL1 cKO mice were compared with WT littermates with intact LanCL1 expression. Formal methods of randomization were not used. All experiments were performed in compliance with the guidelines set forth by Sichuan University and Johns Hopkins University.\n",
      "\n",
      "Disease scoring and behavior analysis\n",
      "\n",
      "Mice were monitored weekly throughout disease progression. Disease status was referenced to body weight changes [29]. Disease onset was defined as the age when mice reached their peak weight before weight began to decline. Disease endpoint was defined as the age at which a mouse could no longer right themselves within 30 s of being put on their side. Disease progression was defined by the duration between the onset and disease endpoint. The early-symptomatic stage (Early-sym) and late-symptomatic (Late-sym) stage was divided based on whether mice had lost 10% of their body weight. Definition of disease course analysis is reported in Fig. 1a.\n",
      "\n",
      "Testing of motor function using a rotarod device (XR1514, Shanghai) began at 8 weeks of age. Each session consisted of three trials on the accelerating rotarod with increasing speed from 4 to 30 rpm in 5 min. The time spent on the rotarod was measured, and for all mice the average of three consecutive trials without prior training was calculated. All behavioral tests were conducted with the experimenter blind to genotypes.\n",
      "\n",
      "Tissue processing, immunofluorescence, and histology\n",
      "\n",
      "Animals were deeply anesthetized with 4% chloral hydrate and were transcardially perfused with phosphate-buffered# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "Onset\n",
      "\n",
      "0%0 ight reduction 1 8 (maximum body weight) Endpoint 0 Pre-sym Early-sym Late-sym 2 120 150 120 150 (days) Pcst-sym IT G934 693/ I P3O P75 P120 P1SO P3O P75 P120 P1SO (kDa) I LanCL1 GAPDH 120 150 420 150 (days) 6934\n",
      "\n",
      "Fig. 1 Dynamic expression of LanCL1 in the spinal cord of ALS mice.\n",
      "\n",
      "nonsignificant, two-way ANOVA followed by Bonferroni post hoc test. Immunoblots (c) and quantifications (d) of LanCL1 in the lumbar spinal cord (L3–L5) of WT and SOD1 G93A mice at presymptomatic stage (P30 and P75), early-symptomatic stage (P120) and late-symptomatic stage (P150). Data represent mean ± SEM, n = 4 mice per genotype per timepoint, ***p < 0.001, n.s. non-significant, two-way ANOVA followed by Bonferroni post hoc test.\n",
      "\n",
      "saline (PBS) followed by 4% PFA in PBS. Lumbar spinal cord (L3–L5) was isolated and post-fixed in 4% PFA overnight at 4 °C. Fixed tissues were then cryoprotected for 48 h in PBS containing 30% sucrose, embedded in O.C.T (TissueTek, Sakura) and sectioned at 20 μm using a freezing microtome (Thermo scientific HM 525). Free-floating sections were mounted onto 2% gelatin-coated slides to be processed for staining.\n",
      "\n",
      "Frozen sections were blocked with 10% normal goat serum/1% BSA/0.3% Triton X-100 in PBS for 1 h at room temperature, and then incubated with combinations of the following primary antibodies: SMI-32 (1:1000, BioLegend, 801701), GFAP (1:1000, Merck Millipore, MAB360), Iba-1 (1:800, Wako, 019-19741). Antibodies were applied in 1% BSA/1% goat serum/0.3% Triton X-100/PBS solution for overnight at 4 °C. Sections were then briefly washed with PBS and incubated for 2 h at room temperature with appropriate combinations of secondary antibodies. After washing, sections were mounted with ProLongGold mounting medium (Life Technologies).\n",
      "\n",
      "For Nissl staining, slides were air-dried overnight. Sections were immersed in 0.1% (w/v) cresyl violet acetate (Nissl stain) (Sigma-Aldrich) using a standard protocol, dehydrated, and coverslipped.\n",
      "\n",
      "For each experimental group or time points, at least three animals were processed and analyzed.\n",
      "\n",
      "Imaging and analysis\n",
      "\n",
      "Tissue sections were visualized using an Olympus BX63 microscope and images captured using a Photometric Sen-Sys Olympus DP70 CCD camera and imaging software (Olympus Corporation, Tokyo, Japan). For each mouse, at least eight images (20×) were taken in the left and right ventral horn (VH) of lumbar spinal cord (L3–L5). Any image contrast and brightness adjustments were made equally between experimental groups.\n",
      "\n",
      "To assess the extent of immunoreactivity for GFAP and Iba-1, an intensity threshold was applied to each image to include the positive labeling while minimizing the inclusion of nonspecific, background staining using ImageJ. The area occupied by the signal (thresholded pixel area) in identically sized regions of ventral spinal cord was determined for each image as previously reported [30]. Values for left and right VH were averaged for each spinal VH and 4–6 sections were analyzed for each mouse (at least three mice per genotype per age group). Results were reported as the fold change with respect to WT (P110).\n",
      "\n",
      "MN quantification\n",
      "\n",
      "The number of MNs were counted in every 10th section of the ventral spinal cord in the series, i.e., at a distance of 200 μm between evaluated slices. The following criteria# 1372\n",
      "\n",
      "H. Tan et al.\n",
      "\n",
      "were used to count Nissl-stained MNs: (1) cells located in the VH, and (2) cells with a maximum diameter of 20 μm or more [31]. Likewise, for SMI-32 staining, only SMI-32+ cells with intact nucleus, ≥20 μm in diameter in the VH were counted [32]. Values for left and right VH that belong to the same cross-section were combined and reported as the MN number per section. The MN numbers were averaged from a minimum of 10 sections and a minimum of 140 neurons per mouse to ensure accurate sampling of heterogeneously distributed MNs. All motoneuron counts were performed in a blinded fashion.\n",
      "\n",
      "TUNEL staining\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 3:\n",
      "\n",
      "Bensimon G, Lacomblez L, Meininger V. A controlled trial of riluzole in amyotrophic lateral sclerosis. ALS/Riluzole Study Group. New Engl J Med. 1994;330:585–91.\n",
      "\n",
      "Yoshino H, Kimura A. Investigation of the therapeutic effects of edaravone, a free radical scavenger, on amyotrophic lateral sclerosis (Phase II study). Amyotroph Lateral Scler. 2006;7:241–5.\n",
      "\n",
      "Ferrante RJ, Browne SE, Shinobu LA, Bowling AC, Baik MJ, MacGarvey U, et al. Evidence of increased oxidative damage in both sporadic and familial amyotrophic lateral sclerosis. J Neurochem. 1997;69:2064–74.\n",
      "\n",
      "Beal MF, Ferrante RJ, Browne SE, Matthews RT, Kowall NW, Brown RH, et al. Increased 3-nitrotyrosine in both sporadic and familial amyotrophic lateral sclerosis. Ann Neurol. 1997;42:644–54.\n",
      "\n",
      "Shibata N, Nagai R, Uchida K, Horiuchi S, Yamada S, Hirano A, et al. Morphological evidence for lipid peroxidation and protein glycoxidation in spinal cords from sporadic amyotrophic lateral sclerosis patients. Brain research. 2001;917:97–104.\n",
      "\n",
      "Bogdanov M, Brown RH, Matson W, Smart R, Hayden D, O’Donnell H, et al. Increased oxidative damage to DNA in ALS patients. Free Radic Biol Med. 2000;29:652–8.\n",
      "\n",
      "Ferrante RJ, Shinobu LA, Schulz JB, Matthews RT, Thomas CE, Kowall NW, et al. Increased 3-nitrotyrosine and oxidative damage in mice with a human copper/zinc superoxide dismutase mutation. Ann Neurol. 1997;42:326–34.\n",
      "\n",
      "Andrus PK, Fleck TJ, Gurney ME, Hall ED. Protein oxidative damage in a transgenic mouse model of familial amyotrophic lateral sclerosis. J Neurochem. 1998;71:2041–8.\n",
      "\n",
      "Liu R, Althaus JS, Ellerbrock BR, Becker DA, Gurney ME. Enhanced oxygen radical production in a transgenic mouse model of familial amyotrophic lateral sclerosis. Ann Neurol. 1998;44:763–70.\n",
      "\n",
      "Barber SC, Mead RJ, Shaw PJ. Oxidative stress in ALS: a mechanism of neurodegeneration and a therapeutic target. Biochim Biophys Acta. 2006;1762:1051–67.\n",
      "\n",
      "Ilieva H, Polymenidou M, Cleveland DW. Non–cell autonomous toxicity in neurodegenerative disorders: ALS and beyond. J Cell Biol. 2009;187:761–72.\n",
      "\n",
      "Alexianu ME, Kozovska M, Appel SH. Immune reactivity in a mouse model of familial ALS correlates with disease progression. Neurology. 2001;57:1282–9.\n",
      "\n",
      "Ferraiuolo L, Kirby J, Grierson AJ, Sendtner M, Shaw PJ. Molecular pathways of motor neuron injury in amyotrophic lateral sclerosis. Nat Rev Neurol. 2011;7:616–30.\n",
      "\n",
      "Kirby J, Ning K, Ferraiuolo L, Heath PR, Ismail A, Kuo SW, et al. Phosphatase and tensin homologue/protein kinase B pathway linked to motor neuron survival in human superoxide dismutase 1-related amyotrophic lateral sclerosis. Brain. 2011;134:506–17.\n",
      "\n",
      "Bauer H, Mayer H, Marchler-Bauer A, Salzer U, Prohaska R. Characterization of p40/GPR69A as a peripheral membrane protein related to the lantibiotic synthetase component C. Biochem Biophys Res Commun. 2000;275:69–74.\n",
      "\n",
      "Chatterjee C, Paul M, Xie L, van der Donk WA. Biosynthesis and mode of action of lantibiotics. Chem Rev. 2005;105:633–84.\n",
      "\n",
      "Mayer H, Salzer U, Breuss J, Ziegler S, Marchler-Bauer A, Prohaska R. Isolation, molecular characterization, and tissue-specific expression of a novel putative G protein-coupled receptor. Biochim Biophys Acta. 1998;1395:301–8.\n",
      "\n",
      "Mayer H, Bauer H, Breuss J, Ziegler S, Prohaska R. Characterization of rat LANCL1, a novel member of the lanthionine synthetase C-like protein family, highly expressed in testis and brain. Gene. 2001;269:73–80.\n",
      "\n",
      "Zhang W, Wang L, Liu Y, Xu J, Zhu G, Cang H, et al. Structure of human lanthionine synthetase C-like protein 1 and its interaction with Eps8 and glutathione. Genes Dev. 2009;23:1387–92.\n",
      "\n",
      "Huang C, Chen M, Pang D, Bi D, Zou Y, Xia X, et al. Developmental and activity-dependent expression of LanCL1 confers antioxidant activity required for neuronal survival. Dev Cell. 2014;30:479–87.\n",
      "\n",
      "Chung CH, Kurien BT, Mehta P, Mhatre M, Mou S, Pye QN, et al. Identification of lanthionine synthase C-like protein-1 as a prominent glutathione binding protein expressed in the mammalian central nervous system. Biochemistry. 2007;46:3262–9.\n",
      "\n",
      "Tronche F, Kellendonk C, Kretz O, Gass P, Anlag K, Orban PC, et al. Disruption of the glucocorticoid receptor gene in the nervous system results in reduced anxiety. Nat Genet. 1999;23:99–103.\n",
      "\n",
      "Boillee S, Yamanaka K, Lobsiger CS, Copeland NG, Jenkins NA, Kassiotis G, et al. Onset and progression in inherited ALS determined by motor neurons and microglia. Science. 2006;312:1389–92.\n",
      "\n",
      "Mironets E, Osei-Owusu P, Bracchi-Ricard V, Fischer R, Owens EA, Ricard J, et al. Soluble TNFalpha signaling within the spinal cord contributes to the development of autonomic dysreflexia and ensuing vascular and immune dysfunction after spinal cord injury. J Neurosci. 2018;38:4146–62.\n",
      "\n",
      "Cai M, Choi SM, Yang EJ. The effects of bee venom acupuncture on the central nervous system and muscle in an animal hSOD1G93A mutant. Toxins. 2015;7:846–58.\n",
      "\n",
      "Guo Y, Zhang Y, Wen D, Duan W, An T, Shi P. et al. The modest impact of transcription factor Nrf2 on the course of disease in an ALS animal model. Lab Investig. 2013;93:825–33.\n",
      "\n",
      "Kaech S, Banker G. Culturing hippocampal neurons. Nat Protoc. 2006;1:2406–15.\n",
      "\n",
      "Feng C, Luo T, Zhang S, Liu K, Zhang Y, Luo Y, et al. Lycopene protects human SHSY5Y neuroblastoma cells against hydrogen peroxide-induced death via inhibition of oxidative stress and mitochondria-associated apoptotic pathways. Molecular medicine reports. 2016;13:4205–14.\n",
      "\n",
      "Ouali Alami N, Schurr C, Olde Heuvel F, Tang L, Li Q, Tasdogan A, et al. NF-kappaB activation in astrocytes drives a stage-specific beneficial neuroimmunological response in ALS. EMBO J. 2018;37:e98697.\n",
      "\n",
      "Miyoshi S, Tezuka T, Arimura S, Tomono T, Okada T, Yamamashi Y. DOK7 gene therapy enhances motor activity and life span in ALS model mice. EMBO Mol Med. 2017;9:880–9.# References\n",
      "\n",
      "Lee JK, Shin JH, Hwang SG, Gwag BJ, McKee AC, Lee J, et al. MST1 functions as a key modulator of neurodegeneration in a mouse model of ALS. PNAS. 2013;110:12066–71.\n",
      "\n",
      "Ludolph AC, Bendotti C, Blaugrund E, Hengerer B, Loffler JP, Martin J, et al. Guidelines for the preclinical in vivo evaluation of pharmacological active drugs for ALS/MND: report on the 142nd ENMC international workshop. Amyotroph Lateral Scler 2007;8: 217–23.\n",
      "\n",
      "Tsang YM, Chiong F, Kuznetsov D, Kasarskis E, Geula C. Motor neurons are rich in non-phosphorylated neurofilaments: cross-species comparison and alterations in ALS. Brain Res. 2000;861:45–58.\n",
      "\n",
      "Nystrom T. Role of oxidative carbonylation in protein quality control and senescence. EMBO J. 2005;24:1311–7.\n",
      "\n",
      "Esterbauer H, Zollner H, Lang J. Metabolism of the lipid peroxidation product 4-hydroxynonenal by isolated hepatocytes and by liver cytosolic fractions. Biochemical J. 1985;228:363–73.\n",
      "\n",
      "Dudek H, Datta SR, Franke TF, Birnbaum MJ, Yao R, Cooper GM, et al. Regulation of neuronal survival by the serine-threonine protein kinase Akt. Science. 1997;275:661–5.\n",
      "\n",
      "Yamaguchi A, Tamatani M, Matsuzaki H, Namikawa K, Kiyama H, Vitek MP, et al. Akt activation protects hippocampal neurons from apoptosis by inhibiting transcriptional activity of p53. J Biol Chem. 2001;276:5256–64.\n",
      "\n",
      "Manning BD, Toker A. AKT/PKB signaling: navigating the network. Cell. 2017;169:381–405.\n",
      "\n",
      "Koh SH, Kwon H, Kim KS, Kim J, Kim MH, Yu HJ, et al. Epigallocatechin gallate prevents oxidative-stress-induced death of mutant Cu/Zn-superoxide dismutase (G93A) motoneuron cells by alteration of cell survival and death signals. Toxicology. 2004;202:213–25.\n",
      "\n",
      "Yin X, Ren M, Jiang H, Cui S, Wang S, Jiang H, et al. Down-regulated AEG-1 together with inhibited PI3K/Akt pathway is associated with reduced viability of motor neurons in an ALS model. Mol Cell Neurosci. 2015;68:303–13.\n",
      "\n",
      "Sadidi M, Lentz SI, Feldman EL. Hydrogen peroxide-induced Akt phosphorylation regulates Bax activation. Biochimie. 2009;91:577–85.\n",
      "\n",
      "Wang X, McCullough KD, Franke TF, Holbrook NJ. Epidermal growth factor receptor-dependent Akt activation by oxidative stress enhances cell survival. J Biol Chem. 2000;275:14624–31.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 4:\n",
      "\n",
      "MN quantification\n",
      "\n",
      "The number of MNs were counted in every 10th section of the ventral spinal cord in the series, i.e., at a distance of 200 μm between evaluated slices. The following criteria# 1372\n",
      "\n",
      "H. Tan et al.\n",
      "\n",
      "were used to count Nissl-stained MNs: (1) cells located in the VH, and (2) cells with a maximum diameter of 20 μm or more [31]. Likewise, for SMI-32 staining, only SMI-32+ cells with intact nucleus, ≥20 μm in diameter in the VH were counted [32]. Values for left and right VH that belong to the same cross-section were combined and reported as the MN number per section. The MN numbers were averaged from a minimum of 10 sections and a minimum of 140 neurons per mouse to ensure accurate sampling of heterogeneously distributed MNs. All motoneuron counts were performed in a blinded fashion.\n",
      "\n",
      "TUNEL staining\n",
      "\n",
      "Mouse spinal cords were fixed as described above for the Nissl-stained sections. Lumbar spinal cord (L3–L5) sections were screened for apoptotic cells by terminal deoxynucleotidyl transferase-mediated dUTP nickend-labeling (TUNEL) method. The Apoptag® Plus In Situ Apoptosis Detection kit (Roche, 11684817910) was used. All the procedures were conducted according to the protocol of the kit and of our previous study [26]. Endogenous peroxidase activity was blocked by incubation in 3% hydrogen peroxide/methanol at room temperature for 10 min. The sections were then incubated with terminal deoxynucleotidyl transferase at 37 °C for 60 min to add the digoxigenin-conjugated dUTP to the 3′-OH ends of fragmented DNA. Converter-POD (anti-fluorescein antibody, Fab fragment from sheep, conjugated with horse-radish peroxidase) was added on sample at 37 °C for 30 min. The sections were visualized using 3,3′-diaminobenzidine hydrochloride (DAB) as chromogen. Positive controls provided in the kit were simultaneously processed.\n",
      "\n",
      "Immunoblotting analysis\n",
      "\n",
      "Lumbar spinal cord (L3–L5) or frontal cortex was homogenized in RIPA lysis buffer containing 50 mM Tris-HCl, (pH 7.4), 150 mM NaCl, 1% Triton X-100, 2% SDS, 1% protease inhibitor Cocktail III (Merck) and phosphatase inhibitor Cocktail III (Biovision). The lysates were centrifuged at 15,000 × g for 20 min at 4 °C to collect supernatants. Protein concentration was quantified using the bicinchoninic acid protein assay kit (Thermo Scientific Pierce). Proteins (20–30 μg) were electrophoresed through SDS-PAGE gels and transferred to a PVDF membrane (Merck Millipore). Membranes were blocked with 5% (w/v) skim milk in TBST for 1 h at room temperature and incubated with the primary antibodies in TBST overnight at 4 °C. The following primary antibodies were used: anti-GAPDH (1:5000, Millipore, MAB374), anti-β-Actin (1:2000, Millipore, 04-1116), anti-4-Hydroxy-2-Nonenal (4-HNE, 1:1000, Abcam, ab48506), anti-dinitrophenol (DNP, 1:1000, Millipore, S7150), anti-phospho-AKT (Ser473) (1:1000, Cell signaling, 4060S), anti-phospho-AKT (Thr308) (1:1000, Cell signaling, 2965S), anti-total AKT (1:1000, Cell signaling, 9272S), anti-BAX (1:1000, Millipore, 06-499), anti-HO-1 (1:1000, Abcam, ab13243) or anti-LanCL1 (1:1000, homemade). Anti-LanCL1 antibody was generated by immunizing rabbits with full-length GST fusion protein and affinity-purified using maltose bind protein (MBP)-tagged LanCL1 recombinant protein. Blots were washed three times in TBST for 10 min each and incubated with secondary antibodies (goat anti-rabbit IgG-HRP or goat anti-mouse IgG-HRP, 1:10,000, Thermo Pierce) for 2 h at room temperature. The proteins were detected using the ECL system and quantified using ImageJ software by taking the mean grey value of bands for target proteins normalized to GAPDH or beta actin level after subtracting background intensity. pAKT bands were normalized to the level of tAKT.\n",
      "\n",
      "RNA extraction and RT-PCR\n",
      "\n",
      "Total RNAs were extracted from tissues using TRIzol reagent (Life Technologies). RNA was subjected to reverse transcription with reverse transcriptase as Manufacturer’s instructions (Thermo). Quantitative real-time PCR was performed using the Bio-Rad CFX96 system, and the relative gene expression was normalized to internal control as GAPDH. Primer sequences for SYBR Green probes of target genes are as follows: LanCL1 forward 5′-CCT TCAGGTGAACCAAGGAA-3′ and reverse 5′-CCAGAG ACCTGCTTGTCCAT-3′; IL-1β forward 5′-CTGGTGT GTGACGTTCCCATTA-3′ and reverse 5′-CCGACAG CACGAGGCTTT-3′; IL-6 forward 5′-TTCCATCCAG TTGCCTTCTTG-3′ and reverse 5′-TTGGGAGTGGTAT CCTCTGTGA-3′; TNFα forward 5′CATCTTCTCAAA ATTCGAGTGACA-3′ and reverse 5′-TGGGAGTAGAC AAGGTACAACCC-3′.\n",
      "\n",
      "Cell cultures and transfection\n",
      "\n",
      "Cortical neurons from E17.5 LanCL1 wild-type (WT) and knockout mice were prepared as described previously [26]. In brief, embryos were dissected and minced well with scissors. The dissociated cells were collected by centrifugation and resuspended in Neurobasal (Gibco) medium supplemented with 5% Horse Serum (Hyclone), 2% GlutaMAX (Gibco), 2% B27 (Gibco), 100 U/ml penicillin and 100 μg/ml streptomycin (Gibco). Cells (1 × 106) were plated in six-well culture plates precoated with polylysine (Sigma, 0.5 mg/ml). After 3 days in culture, 2.5 μM cytosine β-D-arabinofuranoside hydrochloride (Sigma) was added to prevent glial proliferation. Then, the neurons were subsequently maintained with# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "Neurobasal Media (Gibco) containing 1% Horse Serum (Hyclone), 100 U/ml penicillin and 100 μg/ml streptomycin (Gibco), and 1% GlutaMAX (Gibco), 2% B27 (Gibco). These cultures contained >95% neurons and no detectable microglia. Astroglia were generated as described [33].\n",
      "\n",
      "HEK293T and HeLa cells were obtained from the Cell Bank of China Center for Type Culture Collection (Wuhan, China). Both cell lines were authenticated by short tandem repeat profiling. Cells were cultured in Dulbecco’s high glucose modified Eagle medium (DMEM, Life Technologies) supplemented with 10% fetal bovine serum (FBS, Wisent) and 2 mM L-glutamine (Sigma). HEK293T cells were collected 48 h after transfection with Myc-pRK5 vector, LanCL1 (WT), Lipofectamine® LanCL1 (R4A), or LanCL1 (R322A) plasmid using 2000 (Life Technologies), according to the manufacturer’s instructions.\n",
      "\n",
      "Hoechst 33342 staining\n",
      "\n",
      "HeLa cells were transfected with 2 μg of Myc-tagged LanCL1 or Myc-pRK5 plasmid using Lipofectamine® 2000 (Life Technologies). Twenty four hours after transfection, cells were pretreated with either DMSO vehicle, 5 μM BKM120 (Selleck Chemicals, S2247) or 20 μM LY294002 (Selleck Chemicals, S1105) for 30 min prior to addition of 150 μM H2O2. Hoechst 33342 staining was performed 24 h after the H2O2 treatment as previously described [34]. Briefly, cells were washed with PBS, and then incubated with 10 μg/ml Hoechst 33342 dye (Sigma-Aldrich) in the dark for 15 min at room temperature. The apoptotic cells with condensed chromatin and/or fragmented nuclei were visualized and scored under the Olympus BX63 microscope. Three coverslips were used per experimental group, with at least 1000 cells in six random fields being counted. Each experiment was repeated at least three times.\n",
      "\n",
      "Statistical analysis\n",
      "\n",
      "All statistical analyses were performed using GraphPad Prism software version 6 (GraphPad software Inc). No statistical methods were used to predetermine sample sizes, but our sample sizes are similar to or greater than those reported in previous publications [35–37]. Data represent the mean and standard error of the mean (SEM). Unpaired two-tailed Student’s t test was used for the comparison of two means. ANOVA followed by Bonferroni’s or Tukey’s post hoc test were used for the multiple group analysis. Log-rank test was used for disease onset and survival analysis. The significance level for the two-sided analyses was set at P < 0.05.\n",
      "\n",
      "Results\n",
      "\n",
      "CNS-specific LanCL1 transgene expression improves the survival and motor function of ALS mice\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 5:\n",
      "\n",
      "Manning BD, Toker A. AKT/PKB signaling: navigating the network. Cell. 2017;169:381–405.\n",
      "\n",
      "Koh SH, Kwon H, Kim KS, Kim J, Kim MH, Yu HJ, et al. Epigallocatechin gallate prevents oxidative-stress-induced death of mutant Cu/Zn-superoxide dismutase (G93A) motoneuron cells by alteration of cell survival and death signals. Toxicology. 2004;202:213–25.\n",
      "\n",
      "Yin X, Ren M, Jiang H, Cui S, Wang S, Jiang H, et al. Down-regulated AEG-1 together with inhibited PI3K/Akt pathway is associated with reduced viability of motor neurons in an ALS model. Mol Cell Neurosci. 2015;68:303–13.\n",
      "\n",
      "Sadidi M, Lentz SI, Feldman EL. Hydrogen peroxide-induced Akt phosphorylation regulates Bax activation. Biochimie. 2009;91:577–85.\n",
      "\n",
      "Wang X, McCullough KD, Franke TF, Holbrook NJ. Epidermal growth factor receptor-dependent Akt activation by oxidative stress enhances cell survival. J Biol Chem. 2000;275:14624–31.\n",
      "\n",
      "Turner BJ, Talbot K. Transgenics, toxicity and therapeutics in rodent models of mutant SOD1-mediated familial ALS. Prog Neurobiol. 2008;85:94–134.\n",
      "\n",
      "Riboldi G, Nizzardo M, Simone C, Falcone M, Bresolin N, Comi GP, et al. ALS genetic modifiers that increase survival of SOD1 mice and are suitable for therapeutic development. Prog Neurobiol. 2011;95:133-48.\n",
      "\n",
      "Cudkowicz ME, Pastusza KA, Sapp PC, Mathews RK, Leahy J, Pasinelli P, et al. Survival in transgenic ALS mice does not vary with CNS glutathione peroxidase activity. Neurology. 2002;59:729–34.\n",
      "\n",
      "Kang SH, Li Y, Fukaya M, Lorenzini I, Cleveland DW, Ostrow LW, et al. Degeneration and impaired regeneration of gray matter oligodendrocytes in amyotrophic lateral sclerosis. Nat Neurosci. 2013;16:571–9.\n",
      "\n",
      "Frakes AE, Ferraiuolo L, Haidet-Phillips AM, Schmelzer L, Braun L, Miranda CJ, et al. Microglia induce motor neuron death via the classical NF-kappa B pathway in amyotrophic lateral sclerosis. Neuron. 2014;81:1009–23.\n",
      "\n",
      "Qian K, Huang H, Peterson A, Hu B, Maragakis NJ, Ming GL, et al. Sporadic ALS astrocytes induce neuronal degeneration in vivo. Stem Cell Rep. 2017;8:843–55.\n",
      "\n",
      "Dewil M, Lambrechts D, Sciot R, Shaw PJ, Ince PG, Robberecht W, et al. Vascular endothelial growth factor counteracts the loss of phospho-Akt preceding motor neurone degeneration in amyotrophic lateral sclerosis. Neuropathol Appl Neurobiol. 2007;33:499–509.\n",
      "\n",
      "Peviani M, Tortarolo M, Battaglia E, Piva R, Bendotti C. Specific induction of Akt3 in spinal cord motor neurons is neuroprotective in a mouse model of familial amyotrophic lateral sclerosis. Mol Neurobiol. 2014;49:136–48.\n",
      "\n",
      "Zhang J, Wang X, Vikash V, Ye Q, Wu D, Liu Y, et al. ROS and ROS-Mediated Cellular Signaling. Oxid Med Cell Longev. 2016;2016:4350965.\n",
      "\n",
      "Zeng M, van der Donk WA, Chen J. Lanthionine synthetase C-like protein 2 (LanCL2) is a novel regulator of Akt. Mol Biol Cell. 2014;25:3954–61.\n",
      "\n",
      "Hensley K, Venkova K, Christov A. Emerging biological importance of central nervous system lanthionines. Molecules. 2010;15:5581–94.\n",
      "\n",
      "Nada SE, Tulsulkar J, Raghavan A, Hensley K, Shah ZA. A derivative of the CRMP2 binding compound lanthionine ketimine provides neuroprotection in a mouse model of cerebral ischemia. Neurochem Int. 2012;61:1357–63.\n",
      "\n",
      "He C, Zeng M, Dutta D, Koh TH, Chen J, van der Donk WA. LanCL proteins are not Involved in Lanthionine Synthesis in Mammals. Sci Rep. 2017;7:40980.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 6:\n",
      "\n",
      "Cell Death & Differentiation (2020) 27:1369–1382\n",
      "\n",
      "Differentiation\n",
      "\n",
      "A R T I C L E\n",
      "\n",
      "LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "Honglin Tan 1,2 ● Mina Chen 1 ● Dejiang Pang 1 ● Wanxiang Jiang 1 ● Dandan Bi 1,3 ● Chunyu Li 4 ● Chongyangzi Du ● Xiaoqiang Xia 1 ● Huifang Shang 41 ● Wanchun Yang 1 ● Bo Xiao ● Yiyuan Cui 1 ● Chao Huang ● Paul F. Worley 3 1,5\n",
      "\n",
      "Received: 19 November 2018 / Revised: 3 September 2019 / Accepted: 6 September 2019 / Published online: 30 September 2019\n",
      "\n",
      "© The Author(s) 2019. This article is published with open access\n",
      "\n",
      "Abstract\n",
      "\n",
      "Amyotrophic lateral sclerosis (ALS) is a fatal neurodegenerative disease characterized by progressive loss of motor neurons. Improving neuronal survival in ALS remains a significant challenge. Previously, we identified Lanthionine synthetase C-like protein 1 (LanCL1) as a neuronal antioxidant defense gene, the genetic deletion of which causes apoptotic neurodegeneration in the brain. Here, we report in vivo data using the transgenic SOD1G93A mouse model of ALS indicating that CNS-specific expression of LanCL1 transgene extends lifespan, delays disease onset, decelerates symptomatic progression, and improves motor performance of SOD1G93A mice. Conversely, CNS-specific deletion of LanCL1 leads to neurodegenerative phenotypes, including motor neuron loss, neuroinflammation, and oxidative damage. Analysis reveals that LanCL1 is a positive regulator of AKT activity, and LanCL1 overexpression restores the impaired AKT activity in ALS model mice. These findings indicate that LanCL1 regulates neuronal survival through an alternative mechanism, and suggest a new therapeutic target in ALS.\n",
      "\n",
      "These authors contributed equally: Honglin Tan, Mina Chen\n",
      "\n",
      "Introduction\n",
      "\n",
      "Amyotrophic lateral sclerosis (ALS) is a devastating, rapidly progressive neurodegenerative disease characterized by the selective loss of both upper and lower motor neurons (MNs) [1]. Loss of these neurons causes muscle weakness, spasticity, atrophy, paralysis, and premature death [2]. Although the majority of cases are sporadic, ~10% cases are familial (inherited) with mutations in superoxide dismutase 1 (SOD1) being the most extensively studied [3]. Transgenic rodents carrying mutant forms of SOD1 develop a similar, progressive MN disease akin to patients [4]. ALS is more frequent in males [5], and the male patients tend to have an earlier age of onset, but the cause for this gender bias remains enigmatic [6]. Currently, Riluzole and Edaravone are the only two approved drugs for ALS clinical therapy with limited efficacy [7, 8]. Thus, how to promote neuron survival remains an unmet need in ALS treatment.\n",
      "\n",
      "Oxidative stress has been identified as a central feature of ALS pathogenesis. A large body of studies have described evidence of increased oxidative stress in ALS patients [9–12]. Transgenic mouse based on mutant SOD1 recapitulate the oxidative damage to protein, lipid, and DNA observed in the human disease [13–15]. Although oxidative stress being a primary cause of pathogenesis in ALS or a\n",
      "\n",
      "Supplementary information\n",
      "\n",
      "The online version of this article (https://doi.org/10.1038/s41418-019-0422-6) contains supplementary material, which is available to authorized users.\n",
      "\n",
      "Contact Information\n",
      "\n",
      "Mina Chen\n",
      "\n",
      "chenmina2010@scu.edu.cn\n",
      "\n",
      "Paul F. Worley\n",
      "\n",
      "pworley@jhmi.edu\n",
      "\n",
      "Bo Xiao\n",
      "\n",
      "xiaob@sustech.edu.cn\n",
      "\n",
      "Affiliations\n",
      "\n",
      "1 Neuroscience & Metabolism Research, State Key Laboratory of Biotherapy, West China Hospital, Sichuan University and Collaborative Innovation Center, 610041 Chengdu, PR China\n",
      "\n",
      "2 Laboratory of Experimental Animal Disease Model, College of Veterinary Medicine, Sichuan Agricultural University, 611130 Chengdu, PR China\n",
      "\n",
      "3 The Solomon H. Snyder Department of Neuroscience, Johns Hopkins University School of Medicine, Baltimore, MD 21205, USA\n",
      "\n",
      "4 Department of Neurology, West China Hospital, Sichuan University, 610041 Chengdu, PR China\n",
      "\n",
      "5 Department of Biology, Southern University of Science and Technology, 518000 Shenzhen, PR China# H. Tan et al.\n",
      "\n",
      "Downstream consequence of the disease has been debated, it has been proved that oxidative stress is capable of causing considerable damage to MN populations, and can also influence other mechanisms implicated in neurodegeneration in ALS [16].\n",
      "\n",
      "MN cell death, accompanied with astrocytic and microglial cell activation, is the characteristic feature of ALS [17, 18]. Emerging evidence suggests that multiple molecular signaling pathways are involved in MN survival/death in ALS [19]. For example, in SOD1-related ALS in humans, MNs that survive the disease process show upregulation of genes that promote neuronal survival—namely, those encoding the phosphatase and tensin homolog/phosphoinositide 3-kinase/protein kinase B (PTEN/PI3K/AKT) pathway [20]. These data imply that boosting the intrinsic neuroprotective defense mechanisms may represent a promising strategy for ALS treatment.\n",
      "\n",
      "LanCL1 (Lanthionine synthetase C-like protein 1, also known as P40 or GRP69A), is homologous to bacterial lanthionine synthetase C (LanC) family [21], which is involved in the biosynthesis of antimicrobial peptides (lantibiotics) [22]. LanCL1 was identified as a reduced glutathione (GSH)-binding protein and primarily expressed in neural tissues and testis [23–25]. In previous studies, we have revealed that LanCL1 is a critical regulator of neuronal survival during normal postnatal development and in response to oxidative stress [26]. Specifically, loss of LanCL1 leads to neuronal death, oxidative stress, and inflammation in the brain, whereas LanCL1 transgene protects neurons against exogenous peroxide-induced apoptosis [26]. Enzymatic assays reveal that LanCL1 possess ROS scavenging properties like glutathione S-transferases (GSTs) [26]. These findings prompted us to investigate whether LanCL1 could provide neuronal protection in neurodegenerative diseases, such as ALS. Interestingly, LanCL1 protein was found to be increased in the spinal cord of SOD1G93A transgenic mice at presymptomatic stages [27], implicating a possible role of LanCL1 in ALS.\n",
      "\n",
      "Here, we report that LanCL1 protects MNs against degeneration and reduces the severity of disease manifestations in the mouse model of ALS. By crossing LanCL1 conditional transgenic mice with the SOD1G93A mouse model, we found that CNS-specific expression of LanCL1 transgene significantly prolongs lifespan, delays disease onset, decelerates symptomatic progression, and improves motor performance in SOD1G93A mice. Reciprocally, CNS-specific deletion of LanCL1 causes neurodegenerative phenotypes, including MN loss, neuroinflammation, and oxidative damage in spinal cord. We further show that loss of LanCL1 leads to a decrease in AKT phosphorylation, whereas LanCL1 transgene restores AKT phosphorylation and mitigates oxidative stress in ALS mice. Findings suggest that LanCL1 modulates MN survival by scavenging ROS and enhancing AKT activity.\n",
      "\n",
      "Material and methods\n",
      "\n",
      "Animals\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 7:\n",
      "\n",
      "LanCL1 exerts neuronal protective function through mitigating oxidative stress and enhancing AKT activity\n",
      "\n",
      "Deregulation of AKT activity is closely associated with neurodegenerative diseases such as ALS [20, 45, 46]. To determine whether LanCL1 regulated AKT activity plays a role in conferring neuroprotection in ALS mice, we evaluated AKT phosphorylation/activation in the G93A; LanCL1 cKI mice, G93A mice and control littermates. LanCL1 overexpression restored AKT activity which is impaired in G93A mice (Fig. 6a, b). Accordingly, expression of LanCL1 transgene blocked the induction of the proapoptotic Bcl-2 family member BAX (Fig. 6a, b). Consistent with the notion that LanCL1 possess antioxidant function, the oxidative damage in the spinal cord of the G93A mice was ameliorated by LanCL1 transgene. G93A mice exhibited a significant increase in HO-1 (also known as HMOX1) expression, a marker of oxidative stress [13], compared with WT and LanCL1 cKI mice (Fig. 6c, d). However, LanCL1 transgene reduced HO-1 by ~23% in G93A; LanCL1 cKI mice (Fig. 6c, d).\n",
      "\n",
      "LanCL1 regulated AKT activity is relevant to cell survival, because inhibiting activation of PI3K/AKT pathway with the PI3K inhibitor (BKM120 or LY294002) altered survival of HO2-treated 2 cells (Supplementary Fig. 4a, Fig. 6e, f) and the protective effect of LanCL1 overexpression against H2O-induced cell death was compromised by inhibiting AKT activity (Fig. 6e–g). However, LanCL1’s protective effect was not completely abrogated by blocking AKT activation, as more LanCL1-overexpressing# Fig. 5 LanCL1 is a positive regulator of AKT activity.\n",
      "\n",
      "a Spinal cord and cortex lysates from LanCL1 cKO mice and wild-type (WT) littermates were used in western blots to compare the levels of pAKT (T308) and pAKT(S473).\n",
      "\n",
      "b Quantifications showing a reduction in the level of pAKT (T308) in the spinal cord and cortex of LanCL1 cKO mice at P60. Data represent mean ± SEM, spinal cord n = 5 mice per genotype, cortex n = 4 mice per genotype, *p < 0.05, by two-tailed unpaired Student’s t test.\n",
      "\n",
      "c Lysates from DIV 8 and DIV 18 cortical neurons generated from wild-type or LanCL1-deficient mice were used in western blots to compare the levels of pAKT(T308) and pAKT(S473).\n",
      "\n",
      "d Quantifications showing a reduction in the level of pAKT (T308) in cultured LanCL1-deficient cortical neurons at DIV 8 and DIV 18.\n",
      "\n",
      "cko W cKo (kDa) WT DIV8 DIV18 WT LanCL1 pAKT(T308) 2 PAKT(T308) W PAKT(S473) V tAKT 1 2 tAKT B-Actin CTX R-Actin HEK293T cell Ctrl LanCL1 (W) CTX WT cki (kDa) LanCL1 cki CirLanCL1(WT)LanCLI(R4A)LanCLI(R322A)(kDa) LanCLI (R322A) LanCLI n-3 P6O pAKT(S473) 28 PAKT(S473) 20- PAKT(T308) pAKT(T308) tAKT L 2 IAKT 0.5 GAPDH PAKT(S473) PAKT(T308) PAKT(8473) PAKT(T308)\n",
      "\n",
      "cells survived when the cytotoxic effect of HO2 2 was potentiated by inhibiting AKT activity (Fig. 6e–g). This finding argues for an AKT-independent role of LanCL1 in the control of cell survival, which is consistent with its ROS scavenging function as previously described [26]. We are beginning to understand the interplay between LanCL1, oxidative stress and AKT signaling. AKT was activated by HO2 2 treatment in a concentration- and time-dependent manner (Supplementary Fig. 4b–g), consistent with previous studies [47, 48]. When the concentration of H2O2 was low (less than 100 μM), LanCL1 transgene was able to potentiate HO2 2 induced AKT phosphorylation (Supplementary Fig. 4b–d). However, AKT phosphorylation seemed to be already at its maximal level with H2O2 at 150 μM (Supplementary Fig. 4b–g). These results suggest a dynamic relationship among oxidative stress, LanCL1, and AKT activation that needs to be explored in the future studies.\n",
      "\n",
      "Taken together, our findings suggest a model as shown in Fig. 6h, which depicts that LanCL1 could not only reduce oxidative stress as a ROS scavenger, but also function as an AKT activator in the control of cell survival (Fig. 6h).\n",
      "\n",
      "Discussion\n",
      "\n",
      "Our previous work identifies LanCL1 as an antioxidant defense gene which is predominantly expressed in neurons and possesses ROS scavenging properties like GSTs [26]. Here, we found that CNS-specific LanCL1 transgene delays disease onset, decelerates disease progression, and prolongs lifespan in the SOD1G93A mouse model of ALS. These disease-modifying effects of LanCL1 transgene might be a direct result of its role in protecting the MNs in the spinal cord from degeneration in the ALS model, as the age-dependent loss of spinal MNs is significantly delayed. These results demonstrate the neuroprotective effect of LanCL1 transgene in ALS mouse model and support the translational potential of LanCL1.\n",
      "\n",
      "Since the identification of mutations in the SOD1 gene as an important cause of familial ALS [3] and the generation of the SOD1 mouse model of ALS [4], genetic modifications on SOD1 mouse model have been described, mostly focusing on the putative disease mechanisms, such as oxidative stress, excitotoxicity, axonal transport defects,# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "WT (n=4) LanCL1 cKI (n=3) 693A (n=6) G93A LanCL1 cKI (n=4) LarCl 093a (kDa) P110-P140 P125 LanCL1 pAKT (S473) tAKT BAX GAPDH\n",
      "\n",
      "HeLa cells stained with Hoechst 33342\n",
      "\n",
      "DMSO BKM120 (5 μM) Y294002 (20 μM) WT LanCL1 cKI 693A G93A LanCL1 cKI\n",
      "\n",
      "Fig. 6 LanCL1 restores impaired AKT activity and mitigates oxidative stress in ALS mice. Immunoblotting (a) and quantifications (b) of pAKT (S473), pAKT (T308), and BAX in the lumbar spinal cord (L3–L5) of indicated mice. Data represent mean ± SEM; WT n = 4 mice, LanCL1 cKI n = 3 mice, G93A n = 6 mice, G93A; LanCL1 cKI n = 4 mice; p < 0.05, p < 0.01, **p < 0.001; n.s. nonsignificant; one-way ANOVA followed by Tukey’s post hoc test.\n",
      "\n",
      "Immunoblotting (c) and quantifications (d) of HO-1 in the lumbar spinal cord (L3–L5) of indicated mice. Data represent mean ± SEM, n = 4 mice per genotype, p < 0.05, *p < 0.01, one-way ANOVA followed by Tukey’s post hoc test.\n",
      "\n",
      "e HeLa cells were transfected with 2 μg of Myc-pRK5 vector or Myc-tagged LanCL1. Twenty four hours later cells were pretreated with either dimethyl sulfoxide (DMSO), BKM120 (5 μM), or LY294002 (20 μM) for 30 min prior to addition of 150 μM H2O2.\n",
      "\n",
      "Apoptosis assay was performed by Hoechst 33342 staining 24 h after the H2O2 treatment. The percentage of apoptotic HeLa cells with fragmented or condensed nuclei was determined by Hoechst 33342 staining. At least 1000 cells per condition were scored for Hoechst 33342 staining. Data represent mean ± SEM of four independent experiments, p < 0.01, *p < 0.001, two-way ANOVA followed by Bonferroni post hoc test.\n",
      "\n",
      "g Immunoblotting showing that the PI3K inhibitors (BKM120 and LY294002) effectively blocked AKT activation in response to H2O2 treatment in HeLa cells.\n",
      "\n",
      "h Diagram illustrating that LanCL1 regulates cell survival through mitigating oxidative stress and enhancing AKT phosphorylation.# 1380\n",
      "\n",
      "H. Tan et al.\n",
      "\n",
      "mitochondrial dysfunction, and apoptosis [49]. Unfortu- nately, most genetic approaches have failed to extend sur- vival and in some cases have even hastened disease progression in the rodent models of ALS [49, 50]. For example, Genetic elevation or elimination of glutathione peroxidase 1 (GPx1) impacted on neither onset nor survival in SOD1G93A mice [51]. Herein, CNS-specific expression of LanCL1 transgene results in great survival elongation (~15%), delayed disease onset (13–16%), and reduced MN loss (27–34%). It is probable that besides antioxidant defense, other functions of LanCL1 are also important for neuronal protection. Of note, since the Nestin-cre drives LanCL1 overexpression also in glial cells, we cannot rule out the contribution of these glial cells to the overall effects of LanCL1 transgene, as these cells are known to contribute to ALS pathogenesis [52–54].\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 8:\n",
      "\n",
      "G93A LanCL1 cKI WT P110 P135\n",
      "\n",
      "images (c) and quantifications (d) of SMI-32-stained alpha MNs (indicated by white arrows) in matching lumbar ventral horn at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test.# H. Tan et al.\n",
      "\n",
      "WT LanCLI ckl 693A 693A LanCL1 cKi WT P110 LanCLI cki 6934 G934 LanCLi cki n-3 Vh 5 P135 20 Hm [ 8] 3 2 P11O P135\n",
      "\n",
      "Fig. 4 LanCL1 overexpression decreases neuroinflammation in the spinal cord of ALS mice.\n",
      "\n",
      "a Immunostaining of GFAP (green) and DAPI (blue) in matching lumbar ventral horn (VH) at P110 and P135.\n",
      "\n",
      "b Quantification of GFAP immunoreactivity in the lumbar ventral horn of indicated mice at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test respectively.\n",
      "\n",
      "Quantification of MNs with maximum meter ≥ 20 μm showed that in G93A mice, only 47 and 33% of MNs survived at P110 and P135, respectively (Fig. 3a, b). However, in G93A; LanCL1 cKI mice, MN survival was substantially higher; 75 and 66% of the MNs survived at P110 and P135, respectively (Fig. 3a, b).\n",
      "\n",
      "The morphology and survival of MNs was further examined by performing immunostaining with the nonphosphorylated neurofilament marker—SMI-32, which is enriched in alpha MNs of the spinal cord VH [39]. Similarly, G93A mice displayed a progressive reduction in MNs in the lumbar spinal cord, whereas G93A; LanCL1 cKI mice showed significant preservation of MNs at both Early-sym stage (P110) and Late-sym stage (P135) compared with G93A mice (Fig. 3c, d).\n",
      "\n",
      "Spinal MN counts were similar between LanCL1 cKI mice and WT animals (Fig. 3a–d), indicating that no detrimental pathology occurred under CNS-specific overexpression of LanCL1 in vivo. These results indicate that LanCL1 overexpression significantly improves MN survival in the spinal cord of ALS mice.\n",
      "\n",
      "LanCL1 overexpression delays neuroinflammation in ALS mice\n",
      "\n",
      "Neuroinflammation as indicated by astrocytic and microglial activation is a well-documented feature of ALS pathology [1]. We examined whether LanCL1 overexpression mitigates glial pathology of ALS mice. By performing immunostaining with GFAP antibody, we found that the GFAP immunoreactivity was significantly increased in the lumbar VH of G93A mice at Early-sym stage (P110, ~3.5-fold), and further increased over time, being more prominent at Late-sym stage (P135, up to ~5.7-fold) (Fig. 4a, b).\n",
      "\n",
      "However, LanCL1 overexpression decreased the immunoreactivity of GFAP by ~42% (p < 0.05) at P110 and by ~45% (p < 0.05) at P135, respectively (Fig. 4a, b). Similarly, G93A mice showed an age-dependent increase of Iba-1 immunoreactivity in the spinal VH, which was significantly reduced by LanCL1 overexpression in G93A; LanCL1 cKI littermates (Fig. 4c, d).\n",
      "\n",
      "For comparison, the immunoreactivity of GFAP and Iba-1 was not different in WT and LanCL1 cKI mice (Fig. 4a–d). These data indicate# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "that LanCL1 overexpression delays glial pathology in the spinal cord of ALS mice.\n",
      "\n",
      "Deletion of LanCL1 leads to spinal cord pathology\n",
      "\n",
      "Previously, we showed that LanCL1 KO mice develop degenerative phenotypes in the brain [26]. Here, we extend this analysis to include spinal cord of LanCL1 conditional knockout (LanCL1f/f; Nestin Cre+/−, termed LanCL1 cKO) mice. LanCL1 cKO mice exhibited a reduction of MNs by ~18% in the spinal VH at P60-P90, as indicated by Nissl staining (Supplementary Fig. 2a, b). SMI-32 staining revealed a decrease of alpha MNs in LanCL1 cKO spinal VH at the same age (Supplementary Fig. 2a, c). The loss of spinal neurons was associated with increased apoptosis, as we detected a ~3.5-fold increase in the number of TUNEL+ cells in the spinal VH of LanCL1 cKO mice by terminal deoxynucleotidyl transferase-mediated deoxyuridine triphosphate nickend labeling (Supplementary Fig. 2d, e). These results indicate that LanCL1 plays a role in promoting neuronal survival in spinal cord.\n",
      "\n",
      "We next examined whether MN degeneration was accompanied by neuroinflammatory responses. The GFAP and Iba-1 immunoreactivity were significantly increased as early as P60 (Supplementary Fig. 2f–i). Consistently, the mRNA levels of pro-inflammatory cytokines, including IL-1β, IL-6, and TNF-α were significantly upregulated (Supplementary Fig. 2j). Additionally, LanCL1 cKO spinal cord displayed ectopic accumulation of protein carbonyl and lipid peroxidation, as indicated by DNP [40] and 4-HNE [41] western blots (Supplementary Fig. 2k–n). Consistent with neuronal damage, LanCL1 cKO mice develop late stage motor impairment, as indicated by the Rotarod tests of mice at 8–12 month (Supplementary Fig. 2o).\n",
      "\n",
      "LanCL1 positively regulates AKT activity\n",
      "\n",
      "Our previous work demonstrated that LanCL1’s natural function protects neurons through antioxidant defense; deletion of LanCL1 leads to oxidative damage to neurons in the cortex [26]. In the course of examining the mechanisms by which loss of LanCL1 causes neurodegeneration, we found that AKT activation is impaired as a result of LanCL1 deletion. AKT, a serine-threonine kinase, can function to either reduce or prevent cellular destruction from oxidants and plays a critical role in neuronal survival [42, 43]. The activity of AKT is regulated by the phosphorylation of two sites—Thr308 in the catalytic loop by PDK1 and Ser473 in the hydrophobic motif by mTORC2 [44]. We found that loss of LanCL1 significantly decreased the phosphorylation of AKT at Thr308 consistently in the spinal cord and cortex, but to a modest degree at Ser473 and did not reach the level of statistical significance (Fig. 5a, b, Supplementary Fig. 3a). Similar results on AKT phosphorylation status were obtained by the study of cortical neuron cultures derived from WT and LanCL1 KO mice (Fig. 5c, d, Supplementary Fig. 3b).\n",
      "\n",
      "We examined the upstream kinase activity of pAKT(T308)—PDK1, and found that pPDK1(S241) was upregulated in LanCL1 cKO tissues, suggesting that LanCL1 does not regulate AKT phosphorylation through its kinase PDK1 (Supplementary Fig. 3c, d). To examine the role of LanCL1 in the regulation of AKT phosphorylation, we transfected LanCL1 in HEK293T cells and assessed its effect on AKT phosphorylation. LanCL1 transfection elevated AKT phosphorylation at both Thr308 and Ser473 sites (Fig. 5e, f). The effect of LanCL1 on AKT phosphorylation appears to be independent of its GSH-binding capacity because LanCL1 point mutants (R4A and R322A) that failed to bind GSH [25] showed similar effect on AKT phosphorylation as WT LanCL1 (Fig. 5e, f). Consistent with results obtained by in vitro transfection of LanCL1, expression of LanCL1 transgene in neural cells increased AKT phosphorylation at Thr308 and Ser473 in the cortex and spinal cord of LanCL1 cKI mice (Figs. 5g, h and 6a, b). Together, these results indicate that LanCL1 is a positive regulator of AKT activity.\n",
      "\n",
      "LanCL1 exerts neuronal protective function through mitigating oxidative stress and enhancing AKT activity\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 9:\n",
      "\n",
      "H. Tan et al.\n",
      "\n",
      "mitochondrial dysfunction, and apoptosis [49]. Unfortu- nately, most genetic approaches have failed to extend sur- vival and in some cases have even hastened disease progression in the rodent models of ALS [49, 50]. For example, Genetic elevation or elimination of glutathione peroxidase 1 (GPx1) impacted on neither onset nor survival in SOD1G93A mice [51]. Herein, CNS-specific expression of LanCL1 transgene results in great survival elongation (~15%), delayed disease onset (13–16%), and reduced MN loss (27–34%). It is probable that besides antioxidant defense, other functions of LanCL1 are also important for neuronal protection. Of note, since the Nestin-cre drives LanCL1 overexpression also in glial cells, we cannot rule out the contribution of these glial cells to the overall effects of LanCL1 transgene, as these cells are known to contribute to ALS pathogenesis [52–54].\n",
      "\n",
      "The AKT pathway is well known to regulate neuronal survival, and plays a vital role in neurodegenerative diseases, such as ALS [20, 46, 55, 56]. Multiple studies have reported impaired AKT-mediated prosurvival signaling pathway in ALS mouse models and in patients as well [20, 45, 55]. In this study, we identify LanCL1 as a novel activator of AKT kinase. Genetic deletion of LanCL1 reduces AKT phosphorylation in the brain and spinal cord, whereas the expression of LanCL1 transgene in vitro and in vivo elevates AKT phosphorylation. The AKT activity regulated by LanCL1 contributes to cell protection, as inhibiting activation of AKT pathway altered survival of H O 2-treated 2 cells. Here, the neuroprotective effect of LanCL1 transgene in the ALS mouse model could be the combined effect of LanCL1 as a ROS scavenger and an activator of AKT activity. Besides, the relationship between AKT signaling and oxidative stress is complex, since AKT not only modulates oxidative stress but also is affected by ROS activation [48, 57]. Thus, there is an intriguing link between LanCL1, oxidative stress and AKT signaling (Fig. 6h). Additional studies are needed to explore the molecular mechanisms by which LanCL1 regulates the AKT activity. We found that the upstream kinase of AKT–PDK1 activity was upregulated in LanCL1 cKO tissues, suggesting that LanCL1 does not function through activating PDK1. Activating AKT might be a conserved function of LanC-like (LanCL) protein family. LanCL2, a homology of LanCL1, was recently reported to promote AKT activity via direct physical interactions with both the AKT kinase and the substrate [58].\n",
      "\n",
      "Hensley et al. propose that LanCL1 catalyzes the addition of the Cys of glutathione to protein- or peptide-bound dehydroalanine to form lanthionine, analogous to the reaction catalyzed by LanC in bacteria, and then converted to lanthionine ketimine (LK) in the mammalian brain [59]. Surprisingly, a cell-permeable ester derivative of LK (LKE) appears to possess antioxidative stress and antiapoptotic function in G93A NSC-34 MN-like cells and increase survival in SOD1 transgenic mice [59, 60]. Even though the hypothesis that LanCL1 is involved in LK biosynthesis remains controversial at present [61], it is possible that other functional link between LanCL1 and LK may exist and therefore contributes to neuronal protection.\n",
      "\n",
      "LanCL1 is enriched in neurons and dynamically regulated by growth factors, neuronal activity, and oxidative stress [26]. These facts indicate that LanCL1 expression may be dysregulated in neurodegenerative diseases, such as ALS. The endogenous LanCL1 expression is upregulated at late presymptomatic stage (P75) in the spinal cord of ALS mice, which suggests a protective or compensatory function in the time frame prior to protein oxidation, neuroinflammatory acceleration, and MN death, as proposed by Chung CH et al. [27]. However, due to unknown transcriptional or post-transcriptional alternations, LanCL1 expression is progressively decreased as the disease progresses. Thus, further research is needed to focus on the disease progression-dependent expression of LanCL1 in ALS mice.\n",
      "\n",
      "In conclusion, this study demonstrates that LanCL1 confers neuronal protection in a disease setting and establish that LanCL1 is a new therapeutic target for ALS. It also supports the notion that mitigating oxidative stress combined with boosting AKT activity represents a promising therapeutic approach to treating ALS and related neurodegeneration diseases.\n",
      "\n",
      "Acknowledgements\n",
      "\n",
      "This work was supported by the Key Program of National Natural Science Foundation of China (No. 31530042), the National Natural Science Foundation of China (No. 81571195) and the China Postdoctoral Science Foundation funded project (No. 2017M623021).\n",
      "\n",
      "Compliance with ethical standards\n",
      "\n",
      "Conflict of interest The authors declare that they have no conflict of interest.\n",
      "\n",
      "Publisher’s note Springer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations.\n",
      "\n",
      "Open Access This article is licensed under a Creative Commons Attribution 4.0 International License, which permits use, sharing, adaptation, distribution and reproduction in any medium or format, as long as you give appropriate credit to the original author(s) and the source, provide a link to the Creative Commons license, and indicate if changes were made. The images or other third party material in this article are included in the article’s Creative Commons license, unless indicated otherwise in a credit line to the material. If material is not included in the article’s Creative Commons license and your intended use is not permitted by statutory regulation or exceeds the permitted use, you will need to obtain permission directly from the copyright holder. To view a copy of this license, visit http://creativecommons.org/licenses/by/4.0/.# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "References\n",
      "\n",
      "Hardiman O, Al-Chalabi A, Chio A, Corr EM, Logroscino G, Robberecht W, et al. Amyotrophic lateral sclerosis. Nat Rev Dis Primers. 2017;3:17071.\n",
      "\n",
      "Robberecht W, Philips T. The changing scene of amyotrophic lateral sclerosis. Nat Rev Neurosci. 2013;14:248–64.\n",
      "\n",
      "Rosen DR, Siddique T, Patterson D, Figlewicz DA, Sapp P, Hentati A, et al. Mutations in Cu/Zn superoxide dismutase gene are associated with familial amyotrophic lateral sclerosis. Nature. 1993;362:59–62.\n",
      "\n",
      "Gurney ME, Pu H, Chiu AY, Dal Canto MC, Polchow CY, Alexander DD, et al. Motor neuron degeneration in mice that express a human Cu,Zn superoxide dismutase mutation. Science. 1994;264:1772–5.\n",
      "\n",
      "Logroscino G, Traynor BJ, Hardiman O, Chio A, Mitchell D, Swingler RJ, et al. Incidence of amyotrophic lateral sclerosis in Europe. J Neurol Neurosurg Psychiatry. 2010;81:385–90.\n",
      "\n",
      "Heiman-Patterson TD, Deitch JS, Blankenhorn EP, Erwin KL, Perreault MJ, Alexander BK, et al. Background and gender effects on survival in the TgN(SOD1-G93A)1Gur mouse model of ALS. J Neurological Sci. 2005;236:1–7.\n",
      "\n",
      "Bensimon G, Lacomblez L, Meininger V. A controlled trial of riluzole in amyotrophic lateral sclerosis. ALS/Riluzole Study Group. New Engl J Med. 1994;330:585–91.\n",
      "\n",
      "Yoshino H, Kimura A. Investigation of the therapeutic effects of edaravone, a free radical scavenger, on amyotrophic lateral sclerosis (Phase II study). Amyotroph Lateral Scler. 2006;7:241–5.\n",
      "\n",
      "Ferrante RJ, Browne SE, Shinobu LA, Bowling AC, Baik MJ, MacGarvey U, et al. Evidence of increased oxidative damage in both sporadic and familial amyotrophic lateral sclerosis. J Neurochem. 1997;69:2064–74.\n",
      "\n",
      "Beal MF, Ferrante RJ, Browne SE, Matthews RT, Kowall NW, Brown RH, et al. Increased 3-nitrotyrosine in both sporadic and familial amyotrophic lateral sclerosis. Ann Neurol. 1997;42:644–54.\n",
      "\n",
      "Shibata N, Nagai R, Uchida K, Horiuchi S, Yamada S, Hirano A, et al. Morphological evidence for lipid peroxidation and protein glycoxidation in spinal cords from sporadic amyotrophic lateral sclerosis patients. Brain research. 2001;917:97–104.\n"
     ]
    }
   ],
   "source": [
    "relevant_docs = retriever.get_relevant_documents(query)\n",
    "pretty_print_docs(relevant_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 552,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Document 1:\n",
      "\n",
      "Here, we report that LanCL1 protects MNs against degeneration and reduces the severity of disease manifestations in the mouse model of ALS. By crossing LanCL1 conditional transgenic mice with the SOD1G93A mouse model, we found that CNS-specific expression of LanCL1 transgene significantly prolongs lifespan, delays disease onset, decelerates symptomatic progression, and improves motor performance in SOD1G93A mice. Reciprocally, CNS-specific deletion of LanCL1 causes neurodegenerative phenotypes, including MN loss, neuroinflammation, and oxidative damage in spinal cord. We further show that loss of LanCL1 leads to a decrease in AKT phosphorylation, whereas LanCL1 transgene restores AKT phosphorylation and mitigates oxidative stress in ALS mice. Findings suggest that LanCL1 modulates MN survival by scavenging ROS and enhancing AKT activity.\n",
      "\n",
      "Material and methods\n",
      "\n",
      "Animals\n",
      "\n",
      "The generation and characterization of the LanCL1 conditional knockin (termed LanCL1 cKI) mice and LanCL1flox/flox mice have been previously described [26]. SOD1G93A [B6SJL-TgN(SOD1-G93A)1Gur] and Nestin-Cre (C57BL/6) mice were originally derived from the Jackson Laboratory (Bar Harbor, Maine, USA) and maintained on C57BL/6 and 129s4 mixed background. Male mice heterozygous for the human SOD1G93A Cre+/− LanCL1K/K; Nestin transgene were crossbred with female mice, resulting in triple-transgenic mice and control littermates. Littermates were always used as controls. The breeding strategy and the resulting genotypes are reported in Fig. 2a. Mice carrying the floxed allele of LanCL1 were crossed to Nestin-Cre transgenic mice [28] to generate LanCL1 conditional knockout (termed LanCL1 cKO) mice with LanCL1 deficiency in central nervous system (CNS). LanCL1 cKO mice were compared with WT littermates with intact LanCL1 expression. Formal methods of randomization were not used. All experiments were performed in compliance with the guidelines set forth by Sichuan University and Johns Hopkins University.\n",
      "\n",
      "Disease scoring and behavior analysis\n",
      "\n",
      "Mice were monitored weekly throughout disease progression. Disease status was referenced to body weight changes [29]. Disease onset was defined as the age when mice reached their peak weight before weight began to decline. Disease endpoint was defined as the age at which a mouse could no longer right themselves within 30 s of being put on their side. Disease progression was defined by the duration between the onset and disease endpoint. The early-symptomatic stage (Early-sym) and late-symptomatic (Late-sym) stage was divided based on whether mice had lost 10% of their body weight. Definition of disease course analysis is reported in Fig. 1a.\n",
      "\n",
      "Testing of motor function using a rotarod device (XR1514, Shanghai) began at 8 weeks of age. Each session consisted of three trials on the accelerating rotarod with increasing speed from 4 to 30 rpm in 5 min. The time spent on the rotarod was measured, and for all mice the average of three consecutive trials without prior training was calculated. All behavioral tests were conducted with the experimenter blind to genotypes.\n",
      "\n",
      "Tissue processing, immunofluorescence, and histology\n",
      "\n",
      "Animals were deeply anesthetized with 4% chloral hydrate and were transcardially perfused with phosphate-buffered# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "Onset\n",
      "\n",
      "0%0 ight reduction 1 8 (maximum body weight) Endpoint 0 Pre-sym Early-sym Late-sym 2 120 150 120 150 (days) Pcst-sym IT G934 693/ I P3O P75 P120 P1SO P3O P75 P120 P1SO (kDa) I LanCL1 GAPDH 120 150 420 150 (days) 6934\n",
      "\n",
      "Fig. 1 Dynamic expression of LanCL1 in the spinal cord of ALS mice.\n",
      "\n",
      "nonsignificant, two-way ANOVA followed by Bonferroni post hoc test. Immunoblots (c) and quantifications (d) of LanCL1 in the lumbar spinal cord (L3–L5) of WT and SOD1 G93A mice at presymptomatic stage (P30 and P75), early-symptomatic stage (P120) and late-symptomatic stage (P150). Data represent mean ± SEM, n = 4 mice per genotype per timepoint, ***p < 0.001, n.s. non-significant, two-way ANOVA followed by Bonferroni post hoc test.\n",
      "\n",
      "saline (PBS) followed by 4% PFA in PBS. Lumbar spinal cord (L3–L5) was isolated and post-fixed in 4% PFA overnight at 4 °C. Fixed tissues were then cryoprotected for 48 h in PBS containing 30% sucrose, embedded in O.C.T (TissueTek, Sakura) and sectioned at 20 μm using a freezing microtome (Thermo scientific HM 525). Free-floating sections were mounted onto 2% gelatin-coated slides to be processed for staining.\n",
      "\n",
      "Frozen sections were blocked with 10% normal goat serum/1% BSA/0.3% Triton X-100 in PBS for 1 h at room temperature, and then incubated with combinations of the following primary antibodies: SMI-32 (1:1000, BioLegend, 801701), GFAP (1:1000, Merck Millipore, MAB360), Iba-1 (1:800, Wako, 019-19741). Antibodies were applied in 1% BSA/1% goat serum/0.3% Triton X-100/PBS solution for overnight at 4 °C. Sections were then briefly washed with PBS and incubated for 2 h at room temperature with appropriate combinations of secondary antibodies. After washing, sections were mounted with ProLongGold mounting medium (Life Technologies).\n",
      "\n",
      "For Nissl staining, slides were air-dried overnight. Sections were immersed in 0.1% (w/v) cresyl violet acetate (Nissl stain) (Sigma-Aldrich) using a standard protocol, dehydrated, and coverslipped.\n",
      "\n",
      "For each experimental group or time points, at least three animals were processed and analyzed.\n",
      "\n",
      "Imaging and analysis\n",
      "\n",
      "Tissue sections were visualized using an Olympus BX63 microscope and images captured using a Photometric Sen-Sys Olympus DP70 CCD camera and imaging software (Olympus Corporation, Tokyo, Japan). For each mouse, at least eight images (20×) were taken in the left and right ventral horn (VH) of lumbar spinal cord (L3–L5). Any image contrast and brightness adjustments were made equally between experimental groups.\n",
      "\n",
      "To assess the extent of immunoreactivity for GFAP and Iba-1, an intensity threshold was applied to each image to include the positive labeling while minimizing the inclusion of nonspecific, background staining using ImageJ. The area occupied by the signal (thresholded pixel area) in identically sized regions of ventral spinal cord was determined for each image as previously reported [30]. Values for left and right VH were averaged for each spinal VH and 4–6 sections were analyzed for each mouse (at least three mice per genotype per age group). Results were reported as the fold change with respect to WT (P110).\n",
      "\n",
      "MN quantification\n",
      "\n",
      "The number of MNs were counted in every 10th section of the ventral spinal cord in the series, i.e., at a distance of 200 μm between evaluated slices. The following criteria# 1372\n",
      "\n",
      "H. Tan et al.\n",
      "\n",
      "were used to count Nissl-stained MNs: (1) cells located in the VH, and (2) cells with a maximum diameter of 20 μm or more [31]. Likewise, for SMI-32 staining, only SMI-32+ cells with intact nucleus, ≥20 μm in diameter in the VH were counted [32]. Values for left and right VH that belong to the same cross-section were combined and reported as the MN number per section. The MN numbers were averaged from a minimum of 10 sections and a minimum of 140 neurons per mouse to ensure accurate sampling of heterogeneously distributed MNs. All motoneuron counts were performed in a blinded fashion.\n",
      "\n",
      "TUNEL staining\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 2:\n",
      "\n",
      "Statistical analysis\n",
      "\n",
      "All statistical analyses were performed using GraphPad Prism software version 6 (GraphPad software Inc). No statistical methods were used to predetermine sample sizes, but our sample sizes are similar to or greater than those reported in previous publications [35–37]. Data represent the mean and standard error of the mean (SEM). Unpaired two-tailed Student’s t test was used for the comparison of two means. ANOVA followed by Bonferroni’s or Tukey’s post hoc test were used for the multiple group analysis. Log-rank test was used for disease onset and survival analysis. The significance level for the two-sided analyses was set at P < 0.05.\n",
      "\n",
      "Results\n",
      "\n",
      "CNS-specific LanCL1 transgene expression improves the survival and motor function of ALS mice\n",
      "\n",
      "As a first step towards the understanding of the role of LanCL1 in ALS, we characterized the temporal expression patterns of LanCL1 in SOD1G93A mice. Since a decline in body weight is highly correlated with denervation-induced muscle atrophy, weight loss is well established as a simple and objective indicator of the disease course [29]. Thus, we analyzed the disease course based on body weight alterations (Fig. 1a). To evaluate LanCL1 expression at different stages of the disease, both mRNA and protein of lumbar spinal cords from SOD1G93A (G93A) mice and WT littermates were analyzed at the presymptomatic stage (P30 and P75), Early-sym stage (P120) and Late-sym stage (P150). Results showed that both LanCL1 mRNA and protein levels were increased at the late presymptomatic stage (P75) as compared with WT mice (Fig. 1b–d). However, with disease development, the mRNA and protein levels of LanCL1 gradually decreased in SOD1G93A spinal cord compared with WT (Fig. 1b–d).\n",
      "\n",
      "To examine whether LanCL1 confers neuroprotective effect in ALS mice (SOD1G93A), we crossed CNS-specific LanCL1 transgenic mice (LanCL1K/K; Nestin Cre+/−) with ALS mice. Four experimental groups were contrasted as follows: (1) SOD1G93A mice with one or two copies of LanCL1 transgene driven by Nestin-Cre (henceforth referred to as G93A; LanCL1 cKI); (2) SOD1G93A mice without LanCL1 transgene (henceforth referred to as G93A); (3) CNS-specific LanCL1 overexpression without SOD1 G93A (henceforth referred to as LanCL1 cKI); (4) neither LanCL1 transgene nor mutant SOD1 is expressed (henceforth referred to as WT; Fig. 2a).\n",
      "\n",
      "We monitored the survival of G93A and G93A; LanCL1 cKI mice. In males, CNS-specific LanCL1 transgene resulted in a 22 days extension in median survival (141 days for G93A and 163 days for G93A; LanCL1 cKI, respectively; Fig. 2b). Similarly, among females, median survival was increased from 146 days for G93A mice to 168 days for G93A; LanCL1 cKI mice (p < 0.01) (Fig. 2c). Using peak body weight to determine disease onset [38] (Fig. 1a), we found that LanCL1 transgene delayed disease onset by ~13% (p < 0.05) among males, from 105 days for G93A mice to 119 days for G93A; LanCL1 cKI mice (Fig. 2d), and by ~16% (p < 0.05) for females, from 108 days for G93A mice to 125.5 days for G93A; LanCL1 cKI animals (Fig. 2e). In addition, LanCL1 transgene extended the progression of the disease by ~43% (p < 0.01) and ~24% (p < 0.01) for males and females, respectively (Fig. 2f, g).\n",
      "\n",
      "Motor impairment is a prominent feature of ALS mouse model. To examine whether motor deficits in ALS mice was ameliorated by LanCL1 transgene, we analyzed the motor...# H. Tan et al.\n",
      "\n",
      "SOD\n",
      "\n",
      "LanCL\n",
      "\n",
      "Nestin\n",
      "\n",
      "Experimental animals\n",
      "\n",
      "Males Males Males 1C0: 8 2 20 L 89 105 0.01 6934 (n=15) G934 (n=17) 6934 LanCL cKi (n=15) 100 200 100 1C0 150 6934 6934 LanCL cKI\n",
      "\n",
      "Females Females Females 100 2 20 3 0.01 146 38 0.05 125.5 L 6934 (n=21) G934 (n=18) cKi (n=16) G934 LanCL cKi (n=16) 100 200 100 150 G934 6934 anCli cki\n",
      "\n",
      "Males Females 350 350 300 300 250 250 200 200 150 150 100 100 G93A (n=13) C93A (n=12) G934 LanCLI cKl (n=13) 6934 LanCLi cKI (n=11) 11 12 13 14 15 16 17 18 12 13 14 15 16 17 18 Age (weeks) Age (weeks)\n",
      "\n",
      "Function by Rotarod assay. G93A; LanCL1 cKI mice exhibited better performance on Rotarod tests at all time points compared with G93A mice, beginning as early as postnatal day 63 (P63) for females and P70 for males (Fig. 2h, i). Age showed that one or two copies of LanCL1 transgene did not.# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "Fig. 2\n",
      "\n",
      "CNS-specific expression of LanCL1 transgene significantly prolongs the lifespan, delays disease onset, extends disease progression, and improves motor function of ALS mice.\n",
      "\n",
      "a Breeding strategy for the generation of G93A; LanCL1 cKI triple-transgenic mice and control littermates. Kaplan–Meier curves of the survival showing that LanCL1 overexpression increases median survival\n",
      "\n",
      "G93A males (n = 15 mice) G93A; LanCL1 cKI males (n = 15 mice) Median Survival 141 days 163 days\n",
      "\n",
      "b from 146 days for G93A females (n = 21 mice) to 168 days for G93A; LanCL1 cKI females (n = 16 mice). **p < 0.01, by log-rank test. Kaplan–Meier curves of disease onset showing that LanCL1 overexpression delays disease onset\n",
      "\n",
      "G93A males (n = 17 mice) G93A; LanCL1 cKI males (n = 15 mice) Disease Onset 105 days 119 days\n",
      "\n",
      "d from 108 days for G93A females (n = 18 mice) to 125.5 days for G93A; LanCL1 cKI females (n = 16 mice). *p < 0.05, by log-rank test. Mean duration of disease (days from onset to endpoint) showing that LanCL1 overexpression extends disease progression\n",
      "\n",
      "G93A males (n = 16 mice) G93A; LanCL1 cKI males (n = 15 mice) Duration of Disease 30.4 days 43 days\n",
      "\n",
      "f from 37.5 days for G93A females (n = 17 mice) to 46.3 days for G93A; LanCL1 cKI females (n = 18 mice). p < 0.01, by two-tailed unpaired Student’s t test. Rotarod tests showing that LanCL1 overexpression improves motor function of male (h) and female (i) G93A mice. Data represent mean ± SEM, n = 11–13 mice per sex per genotype, p < 0.05, p < 0.001, by Mann–Whitney U-tests\n",
      "\n",
      "Fig. 3\n",
      "\n",
      "LanCL1 overexpression rescues MN loss in the spinal cord of ALS mice. Representative images (a) and quantifications (b) of Nissl-stained MNs (indicated by white arrows) in matching lumbar ventral horn (VH) at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test. Representative make a significant difference. Western blotting showed that a single copy of LanCL1 transgene substantially increased LanCL1 protein level in the spinal cord relative to WT tissue, and two copies of LanCL1 did not produce additional increase in LanCL1 protein (Supplementary Fig. 1a, b).\n",
      "\n",
      "LanCL1 overexpression delays MN loss in ALS mice\n",
      "\n",
      "The improved performance in these behavioral tests suggests that the MN degeneration in the SOD1 mutant mice might be delayed by LanCL1 transgene. To examine this notion, we compared MN survival in G93A; LanCL1 cKI and G93A mice, relative to that in WT littermates by performing Nissl staining in the spinal VH of these comparable mice at Early-sym stage (P110) and Late-sym stage (P135),\n",
      "\n",
      "G93A LanCL1 cKI WT P110 P135\n",
      "\n",
      "images (c) and quantifications (d) of SMI-32-stained alpha MNs (indicated by white arrows) in matching lumbar ventral horn at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test.# H. Tan et al.\n",
      "\n",
      "WT LanCLI ckl 693A 693A LanCL1 cKi WT P110 LanCLI cki 6934 G934 LanCLi cki n-3 Vh 5 P135 20 Hm [ 8] 3 2 P11O P135\n",
      "\n",
      "Fig. 4 LanCL1 overexpression decreases neuroinflammation in the spinal cord of ALS mice.\n",
      "\n",
      "a Immunostaining of GFAP (green) and DAPI (blue) in matching lumbar ventral horn (VH) at P110 and P135.\n",
      "\n",
      "b Quantification of GFAP immunoreactivity in the lumbar ventral horn of indicated mice at P110 and P135. Data represent mean ± SEM, n = 3 mice per genotype per age group, p < 0.05, *p < 0.01, two-way ANOVA followed by Tukey’s post hoc test respectively.\n",
      "----------------------------------------------------------------------------------------------------\n",
      "Document 3:\n",
      "\n",
      "MN quantification\n",
      "\n",
      "The number of MNs were counted in every 10th section of the ventral spinal cord in the series, i.e., at a distance of 200 μm between evaluated slices. The following criteria# 1372\n",
      "\n",
      "H. Tan et al.\n",
      "\n",
      "were used to count Nissl-stained MNs: (1) cells located in the VH, and (2) cells with a maximum diameter of 20 μm or more [31]. Likewise, for SMI-32 staining, only SMI-32+ cells with intact nucleus, ≥20 μm in diameter in the VH were counted [32]. Values for left and right VH that belong to the same cross-section were combined and reported as the MN number per section. The MN numbers were averaged from a minimum of 10 sections and a minimum of 140 neurons per mouse to ensure accurate sampling of heterogeneously distributed MNs. All motoneuron counts were performed in a blinded fashion.\n",
      "\n",
      "TUNEL staining\n",
      "\n",
      "Mouse spinal cords were fixed as described above for the Nissl-stained sections. Lumbar spinal cord (L3–L5) sections were screened for apoptotic cells by terminal deoxynucleotidyl transferase-mediated dUTP nickend-labeling (TUNEL) method. The Apoptag® Plus In Situ Apoptosis Detection kit (Roche, 11684817910) was used. All the procedures were conducted according to the protocol of the kit and of our previous study [26]. Endogenous peroxidase activity was blocked by incubation in 3% hydrogen peroxide/methanol at room temperature for 10 min. The sections were then incubated with terminal deoxynucleotidyl transferase at 37 °C for 60 min to add the digoxigenin-conjugated dUTP to the 3′-OH ends of fragmented DNA. Converter-POD (anti-fluorescein antibody, Fab fragment from sheep, conjugated with horse-radish peroxidase) was added on sample at 37 °C for 30 min. The sections were visualized using 3,3′-diaminobenzidine hydrochloride (DAB) as chromogen. Positive controls provided in the kit were simultaneously processed.\n",
      "\n",
      "Immunoblotting analysis\n",
      "\n",
      "Lumbar spinal cord (L3–L5) or frontal cortex was homogenized in RIPA lysis buffer containing 50 mM Tris-HCl, (pH 7.4), 150 mM NaCl, 1% Triton X-100, 2% SDS, 1% protease inhibitor Cocktail III (Merck) and phosphatase inhibitor Cocktail III (Biovision). The lysates were centrifuged at 15,000 × g for 20 min at 4 °C to collect supernatants. Protein concentration was quantified using the bicinchoninic acid protein assay kit (Thermo Scientific Pierce). Proteins (20–30 μg) were electrophoresed through SDS-PAGE gels and transferred to a PVDF membrane (Merck Millipore). Membranes were blocked with 5% (w/v) skim milk in TBST for 1 h at room temperature and incubated with the primary antibodies in TBST overnight at 4 °C. The following primary antibodies were used: anti-GAPDH (1:5000, Millipore, MAB374), anti-β-Actin (1:2000, Millipore, 04-1116), anti-4-Hydroxy-2-Nonenal (4-HNE, 1:1000, Abcam, ab48506), anti-dinitrophenol (DNP, 1:1000, Millipore, S7150), anti-phospho-AKT (Ser473) (1:1000, Cell signaling, 4060S), anti-phospho-AKT (Thr308) (1:1000, Cell signaling, 2965S), anti-total AKT (1:1000, Cell signaling, 9272S), anti-BAX (1:1000, Millipore, 06-499), anti-HO-1 (1:1000, Abcam, ab13243) or anti-LanCL1 (1:1000, homemade). Anti-LanCL1 antibody was generated by immunizing rabbits with full-length GST fusion protein and affinity-purified using maltose bind protein (MBP)-tagged LanCL1 recombinant protein. Blots were washed three times in TBST for 10 min each and incubated with secondary antibodies (goat anti-rabbit IgG-HRP or goat anti-mouse IgG-HRP, 1:10,000, Thermo Pierce) for 2 h at room temperature. The proteins were detected using the ECL system and quantified using ImageJ software by taking the mean grey value of bands for target proteins normalized to GAPDH or beta actin level after subtracting background intensity. pAKT bands were normalized to the level of tAKT.\n",
      "\n",
      "RNA extraction and RT-PCR\n",
      "\n",
      "Total RNAs were extracted from tissues using TRIzol reagent (Life Technologies). RNA was subjected to reverse transcription with reverse transcriptase as Manufacturer’s instructions (Thermo). Quantitative real-time PCR was performed using the Bio-Rad CFX96 system, and the relative gene expression was normalized to internal control as GAPDH. Primer sequences for SYBR Green probes of target genes are as follows: LanCL1 forward 5′-CCT TCAGGTGAACCAAGGAA-3′ and reverse 5′-CCAGAG ACCTGCTTGTCCAT-3′; IL-1β forward 5′-CTGGTGT GTGACGTTCCCATTA-3′ and reverse 5′-CCGACAG CACGAGGCTTT-3′; IL-6 forward 5′-TTCCATCCAG TTGCCTTCTTG-3′ and reverse 5′-TTGGGAGTGGTAT CCTCTGTGA-3′; TNFα forward 5′CATCTTCTCAAA ATTCGAGTGACA-3′ and reverse 5′-TGGGAGTAGAC AAGGTACAACCC-3′.\n",
      "\n",
      "Cell cultures and transfection\n",
      "\n",
      "Cortical neurons from E17.5 LanCL1 wild-type (WT) and knockout mice were prepared as described previously [26]. In brief, embryos were dissected and minced well with scissors. The dissociated cells were collected by centrifugation and resuspended in Neurobasal (Gibco) medium supplemented with 5% Horse Serum (Hyclone), 2% GlutaMAX (Gibco), 2% B27 (Gibco), 100 U/ml penicillin and 100 μg/ml streptomycin (Gibco). Cells (1 × 106) were plated in six-well culture plates precoated with polylysine (Sigma, 0.5 mg/ml). After 3 days in culture, 2.5 μM cytosine β-D-arabinofuranoside hydrochloride (Sigma) was added to prevent glial proliferation. Then, the neurons were subsequently maintained with# LanCL1 promotes motor neuron survival and extends the lifespan of amyotrophic lateral sclerosis mice\n",
      "\n",
      "Neurobasal Media (Gibco) containing 1% Horse Serum (Hyclone), 100 U/ml penicillin and 100 μg/ml streptomycin (Gibco), and 1% GlutaMAX (Gibco), 2% B27 (Gibco). These cultures contained >95% neurons and no detectable microglia. Astroglia were generated as described [33].\n",
      "\n",
      "HEK293T and HeLa cells were obtained from the Cell Bank of China Center for Type Culture Collection (Wuhan, China). Both cell lines were authenticated by short tandem repeat profiling. Cells were cultured in Dulbecco’s high glucose modified Eagle medium (DMEM, Life Technologies) supplemented with 10% fetal bovine serum (FBS, Wisent) and 2 mM L-glutamine (Sigma). HEK293T cells were collected 48 h after transfection with Myc-pRK5 vector, LanCL1 (WT), Lipofectamine® LanCL1 (R4A), or LanCL1 (R322A) plasmid using 2000 (Life Technologies), according to the manufacturer’s instructions.\n",
      "\n",
      "Hoechst 33342 staining\n",
      "\n",
      "HeLa cells were transfected with 2 μg of Myc-tagged LanCL1 or Myc-pRK5 plasmid using Lipofectamine® 2000 (Life Technologies). Twenty four hours after transfection, cells were pretreated with either DMSO vehicle, 5 μM BKM120 (Selleck Chemicals, S2247) or 20 μM LY294002 (Selleck Chemicals, S1105) for 30 min prior to addition of 150 μM H2O2. Hoechst 33342 staining was performed 24 h after the H2O2 treatment as previously described [34]. Briefly, cells were washed with PBS, and then incubated with 10 μg/ml Hoechst 33342 dye (Sigma-Aldrich) in the dark for 15 min at room temperature. The apoptotic cells with condensed chromatin and/or fragmented nuclei were visualized and scored under the Olympus BX63 microscope. Three coverslips were used per experimental group, with at least 1000 cells in six random fields being counted. Each experiment was repeated at least three times.\n",
      "\n",
      "Statistical analysis\n",
      "\n",
      "All statistical analyses were performed using GraphPad Prism software version 6 (GraphPad software Inc). No statistical methods were used to predetermine sample sizes, but our sample sizes are similar to or greater than those reported in previous publications [35–37]. Data represent the mean and standard error of the mean (SEM). Unpaired two-tailed Student’s t test was used for the comparison of two means. ANOVA followed by Bonferroni’s or Tukey’s post hoc test were used for the multiple group analysis. Log-rank test was used for disease onset and survival analysis. The significance level for the two-sided analyses was set at P < 0.05.\n",
      "\n",
      "Results\n",
      "\n",
      "CNS-specific LanCL1 transgene expression improves the survival and motor function of ALS mice\n"
     ]
    }
   ],
   "source": [
    "compressor = CohereRerank(model='rerank-english-v3.0',\n",
    "                          cohere_api_key=COHERE_TOKEN)\n",
    "compression_retriever = ContextualCompressionRetriever(\n",
    "    base_compressor=compressor, base_retriever=retriever\n",
    ")\n",
    "compressed_docs = compression_retriever.get_relevant_documents(query)\n",
    "pretty_print_docs(compressed_docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = ChatMistralAI(model_name=\"mistral-large-latest\",temperature=0,api_key=MISTRAL_TOKEN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 556,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Animal(BaseModel):\n",
    "    species: str = Field(description=\"Species of the animal\")\n",
    "    strain: str = Field(description=\"Strain of the animal\")\n",
    "    group: str = Field(description=\"Control or experiment group\")\n",
    "    gender: str = Field(description=\"Sex of the animal\")\n",
    "    n_treatment: Optional[int] = Field(description='Number of animals in this group')\n",
    "    n_control: Optional[int] = Field(description='Number of animals in control group')\n",
    "\n",
    "class AnimalList(BaseModel):\n",
    "    animals: list[Animal]\n",
    "parser = PydanticOutputParser(pydantic_object=AnimalList)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 557,
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt_template = \"\"\"\n",
    "You are an assistant. Use the following information to answer the question very shortly.\n",
    "Identify ONLY main experimental subject groups and describe them (for example female and male groups, groups with different treatments and e.t.c).\n",
    "Control, wild types and any other groups that are not key subjects of the experiment use only for fill n_control field\n",
    "Give an aswer in proper JSON format using double quotes around keys and values format\n",
    "For example: {{\"animals\":[{{\"species\":\"animal_species1\",\n",
    "         \"strain\":\"animal_strain1\", # ONLY strain name of the animal group\n",
    "         \"group\":\"experiment1\",# Name of the group for example Rapa, KO \"ABC\" gene and e.t.c\n",
    "         \"gender\":\"male\",\n",
    "         \"n_treatment\":25,# Number of animals in this group\n",
    "         \"n_control\":40 # Number of animals in control group\n",
    "         }},\n",
    "         {{\"species\":\"animal_species2\",\n",
    "         \"strain\":\"animal_strain2\",# ONLY strain name of the animal group ONLY name\n",
    "         \"group\":\"experiment2\",# Name of the group for example Rapa, KO ABC gene and e.t.c\n",
    "         \"sex\":\"female\",\n",
    "         \"n_treatment\":25,# Number of animals in this group\n",
    "         \"n_control\":40 # Number of animals in control group\n",
    "         }}]}}\n",
    "Context: {context}\n",
    "Question: {question}\n",
    "Answer:\n",
    "\"\"\"\n",
    "\n",
    "# Create a PromptTemplate instance, note the use of both 'context' and 'query'\n",
    "prompt = PromptTemplate(\n",
    "    template=prompt_template,\n",
    "    input_variables=[\"query\", \"context\"],\n",
    "    partial_variables={\"format_instructions\": parser.model_json_schema()},\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 558,
   "metadata": {},
   "outputs": [],
   "source": [
    "qa = RetrievalQA.from_chain_type(llm=llm,\n",
    "                                 chain_type=\"stuff\",\n",
    "                                 retriever=compression_retriever,\n",
    "                                 chain_type_kwargs={\n",
    "                                     \"prompt\": prompt,\n",
    "                                     \"document_variable_name\": \"context\"\n",
    "                                 })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 559,
   "metadata": {},
   "outputs": [
    {
     "ename": "NotFoundError",
     "evalue": "Error code: 404 - {'error': {'message': 'The model `gpt-4o` does not exist or you do not have access to it.', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_found'}}",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNotFoundError\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[559], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43mqa\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mquery\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/_api/deprecation.py:182\u001b[0m, in \u001b[0;36mdeprecated.<locals>.deprecate.<locals>.warning_emitting_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    180\u001b[0m     warned \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    181\u001b[0m     emit_warning()\n\u001b[0;32m--> 182\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:606\u001b[0m, in \u001b[0;36mChain.run\u001b[0;34m(self, callbacks, tags, metadata, *args, **kwargs)\u001b[0m\n\u001b[1;32m    604\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    605\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m`run` supports only one positional argument.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 606\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtags\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmetadata\u001b[49m\u001b[43m)\u001b[49m[\n\u001b[1;32m    607\u001b[0m         _output_key\n\u001b[1;32m    608\u001b[0m     ]\n\u001b[1;32m    610\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwargs \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m args:\n\u001b[1;32m    611\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m(kwargs, callbacks\u001b[38;5;241m=\u001b[39mcallbacks, tags\u001b[38;5;241m=\u001b[39mtags, metadata\u001b[38;5;241m=\u001b[39mmetadata)[\n\u001b[1;32m    612\u001b[0m         _output_key\n\u001b[1;32m    613\u001b[0m     ]\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/_api/deprecation.py:182\u001b[0m, in \u001b[0;36mdeprecated.<locals>.deprecate.<locals>.warning_emitting_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    180\u001b[0m     warned \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    181\u001b[0m     emit_warning()\n\u001b[0;32m--> 182\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:389\u001b[0m, in \u001b[0;36mChain.__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, run_name, include_run_info)\u001b[0m\n\u001b[1;32m    357\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Execute the chain.\u001b[39;00m\n\u001b[1;32m    358\u001b[0m \n\u001b[1;32m    359\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    380\u001b[0m \u001b[38;5;124;03m        `Chain.output_keys`.\u001b[39;00m\n\u001b[1;32m    381\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    382\u001b[0m config \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    383\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcallbacks\u001b[39m\u001b[38;5;124m\"\u001b[39m: callbacks,\n\u001b[1;32m    384\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtags\u001b[39m\u001b[38;5;124m\"\u001b[39m: tags,\n\u001b[1;32m    385\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m\"\u001b[39m: metadata,\n\u001b[1;32m    386\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_name\u001b[39m\u001b[38;5;124m\"\u001b[39m: run_name,\n\u001b[1;32m    387\u001b[0m }\n\u001b[0;32m--> 389\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    390\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    391\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcast\u001b[49m\u001b[43m(\u001b[49m\u001b[43mRunnableConfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43mk\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mitems\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    392\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_only_outputs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_only_outputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    393\u001b[0m \u001b[43m    \u001b[49m\u001b[43minclude_run_info\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minclude_run_info\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    394\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:170\u001b[0m, in \u001b[0;36mChain.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    169\u001b[0m     run_manager\u001b[38;5;241m.\u001b[39mon_chain_error(e)\n\u001b[0;32m--> 170\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    171\u001b[0m run_manager\u001b[38;5;241m.\u001b[39mon_chain_end(outputs)\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m include_run_info:\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:160\u001b[0m, in \u001b[0;36mChain.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_inputs(inputs)\n\u001b[1;32m    159\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m--> 160\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_manager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    161\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m new_arg_supported\n\u001b[1;32m    162\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_call(inputs)\n\u001b[1;32m    163\u001b[0m     )\n\u001b[1;32m    165\u001b[0m     final_outputs: Dict[\u001b[38;5;28mstr\u001b[39m, Any] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprep_outputs(\n\u001b[1;32m    166\u001b[0m         inputs, outputs, return_only_outputs\n\u001b[1;32m    167\u001b[0m     )\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/retrieval_qa/base.py:154\u001b[0m, in \u001b[0;36mBaseRetrievalQA._call\u001b[0;34m(self, inputs, run_manager)\u001b[0m\n\u001b[1;32m    152\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    153\u001b[0m     docs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_docs(question)  \u001b[38;5;66;03m# type: ignore[call-arg]\u001b[39;00m\n\u001b[0;32m--> 154\u001b[0m answer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcombine_documents_chain\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    155\u001b[0m \u001b[43m    \u001b[49m\u001b[43minput_documents\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdocs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mquestion\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquestion\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_run_manager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_child\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    156\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    158\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturn_source_documents:\n\u001b[1;32m    159\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m {\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moutput_key: answer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msource_documents\u001b[39m\u001b[38;5;124m\"\u001b[39m: docs}\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/_api/deprecation.py:182\u001b[0m, in \u001b[0;36mdeprecated.<locals>.deprecate.<locals>.warning_emitting_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    180\u001b[0m     warned \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    181\u001b[0m     emit_warning()\n\u001b[0;32m--> 182\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:611\u001b[0m, in \u001b[0;36mChain.run\u001b[0;34m(self, callbacks, tags, metadata, *args, **kwargs)\u001b[0m\n\u001b[1;32m    606\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m(args[\u001b[38;5;241m0\u001b[39m], callbacks\u001b[38;5;241m=\u001b[39mcallbacks, tags\u001b[38;5;241m=\u001b[39mtags, metadata\u001b[38;5;241m=\u001b[39mmetadata)[\n\u001b[1;32m    607\u001b[0m         _output_key\n\u001b[1;32m    608\u001b[0m     ]\n\u001b[1;32m    610\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwargs \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m args:\n\u001b[0;32m--> 611\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtags\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmetadata\u001b[49m\u001b[43m)\u001b[49m[\n\u001b[1;32m    612\u001b[0m         _output_key\n\u001b[1;32m    613\u001b[0m     ]\n\u001b[1;32m    615\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m kwargs \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m args:\n\u001b[1;32m    616\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    617\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m`run` supported with either positional arguments or keyword arguments,\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    618\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m but none were provided.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    619\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/_api/deprecation.py:182\u001b[0m, in \u001b[0;36mdeprecated.<locals>.deprecate.<locals>.warning_emitting_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    180\u001b[0m     warned \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    181\u001b[0m     emit_warning()\n\u001b[0;32m--> 182\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:389\u001b[0m, in \u001b[0;36mChain.__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, run_name, include_run_info)\u001b[0m\n\u001b[1;32m    357\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Execute the chain.\u001b[39;00m\n\u001b[1;32m    358\u001b[0m \n\u001b[1;32m    359\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    380\u001b[0m \u001b[38;5;124;03m        `Chain.output_keys`.\u001b[39;00m\n\u001b[1;32m    381\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    382\u001b[0m config \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    383\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcallbacks\u001b[39m\u001b[38;5;124m\"\u001b[39m: callbacks,\n\u001b[1;32m    384\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtags\u001b[39m\u001b[38;5;124m\"\u001b[39m: tags,\n\u001b[1;32m    385\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m\"\u001b[39m: metadata,\n\u001b[1;32m    386\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_name\u001b[39m\u001b[38;5;124m\"\u001b[39m: run_name,\n\u001b[1;32m    387\u001b[0m }\n\u001b[0;32m--> 389\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    390\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    391\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcast\u001b[49m\u001b[43m(\u001b[49m\u001b[43mRunnableConfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43mk\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mitems\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    392\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_only_outputs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_only_outputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    393\u001b[0m \u001b[43m    \u001b[49m\u001b[43minclude_run_info\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minclude_run_info\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    394\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:170\u001b[0m, in \u001b[0;36mChain.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    169\u001b[0m     run_manager\u001b[38;5;241m.\u001b[39mon_chain_error(e)\n\u001b[0;32m--> 170\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    171\u001b[0m run_manager\u001b[38;5;241m.\u001b[39mon_chain_end(outputs)\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m include_run_info:\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:160\u001b[0m, in \u001b[0;36mChain.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_inputs(inputs)\n\u001b[1;32m    159\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m--> 160\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_manager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    161\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m new_arg_supported\n\u001b[1;32m    162\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_call(inputs)\n\u001b[1;32m    163\u001b[0m     )\n\u001b[1;32m    165\u001b[0m     final_outputs: Dict[\u001b[38;5;28mstr\u001b[39m, Any] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprep_outputs(\n\u001b[1;32m    166\u001b[0m         inputs, outputs, return_only_outputs\n\u001b[1;32m    167\u001b[0m     )\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/combine_documents/base.py:138\u001b[0m, in \u001b[0;36mBaseCombineDocumentsChain._call\u001b[0;34m(self, inputs, run_manager)\u001b[0m\n\u001b[1;32m    136\u001b[0m \u001b[38;5;66;03m# Other keys are assumed to be needed for LLM prediction\u001b[39;00m\n\u001b[1;32m    137\u001b[0m other_keys \u001b[38;5;241m=\u001b[39m {k: v \u001b[38;5;28;01mfor\u001b[39;00m k, v \u001b[38;5;129;01min\u001b[39;00m inputs\u001b[38;5;241m.\u001b[39mitems() \u001b[38;5;28;01mif\u001b[39;00m k \u001b[38;5;241m!=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minput_key}\n\u001b[0;32m--> 138\u001b[0m output, extra_return_dict \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcombine_docs\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    139\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdocs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_run_manager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_child\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mother_keys\u001b[49m\n\u001b[1;32m    140\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    141\u001b[0m extra_return_dict[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moutput_key] \u001b[38;5;241m=\u001b[39m output\n\u001b[1;32m    142\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m extra_return_dict\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/combine_documents/stuff.py:259\u001b[0m, in \u001b[0;36mStuffDocumentsChain.combine_docs\u001b[0;34m(self, docs, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    257\u001b[0m inputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_inputs(docs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    258\u001b[0m \u001b[38;5;66;03m# Call predict on the LLM.\u001b[39;00m\n\u001b[0;32m--> 259\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mllm_chain\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43minputs\u001b[49m\u001b[43m)\u001b[49m, {}\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/llm.py:318\u001b[0m, in \u001b[0;36mLLMChain.predict\u001b[0;34m(self, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    303\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpredict\u001b[39m(\u001b[38;5;28mself\u001b[39m, callbacks: Callbacks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mstr\u001b[39m:\n\u001b[1;32m    304\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Format prompt with kwargs and pass to LLM.\u001b[39;00m\n\u001b[1;32m    305\u001b[0m \n\u001b[1;32m    306\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[38;5;124;03m            completion = llm.predict(adjective=\"funny\")\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 318\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m)\u001b[49m[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moutput_key]\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/_api/deprecation.py:182\u001b[0m, in \u001b[0;36mdeprecated.<locals>.deprecate.<locals>.warning_emitting_wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    180\u001b[0m     warned \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    181\u001b[0m     emit_warning()\n\u001b[0;32m--> 182\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mwrapped\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:389\u001b[0m, in \u001b[0;36mChain.__call__\u001b[0;34m(self, inputs, return_only_outputs, callbacks, tags, metadata, run_name, include_run_info)\u001b[0m\n\u001b[1;32m    357\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Execute the chain.\u001b[39;00m\n\u001b[1;32m    358\u001b[0m \n\u001b[1;32m    359\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    380\u001b[0m \u001b[38;5;124;03m        `Chain.output_keys`.\u001b[39;00m\n\u001b[1;32m    381\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    382\u001b[0m config \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    383\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcallbacks\u001b[39m\u001b[38;5;124m\"\u001b[39m: callbacks,\n\u001b[1;32m    384\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtags\u001b[39m\u001b[38;5;124m\"\u001b[39m: tags,\n\u001b[1;32m    385\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m\"\u001b[39m: metadata,\n\u001b[1;32m    386\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_name\u001b[39m\u001b[38;5;124m\"\u001b[39m: run_name,\n\u001b[1;32m    387\u001b[0m }\n\u001b[0;32m--> 389\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minvoke\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    390\u001b[0m \u001b[43m    \u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    391\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcast\u001b[49m\u001b[43m(\u001b[49m\u001b[43mRunnableConfig\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43mk\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mitems\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    392\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_only_outputs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_only_outputs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    393\u001b[0m \u001b[43m    \u001b[49m\u001b[43minclude_run_info\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minclude_run_info\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    394\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:170\u001b[0m, in \u001b[0;36mChain.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    169\u001b[0m     run_manager\u001b[38;5;241m.\u001b[39mon_chain_error(e)\n\u001b[0;32m--> 170\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    171\u001b[0m run_manager\u001b[38;5;241m.\u001b[39mon_chain_end(outputs)\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m include_run_info:\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/base.py:160\u001b[0m, in \u001b[0;36mChain.invoke\u001b[0;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_inputs(inputs)\n\u001b[1;32m    159\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m--> 160\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43minputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_manager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    161\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m new_arg_supported\n\u001b[1;32m    162\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_call(inputs)\n\u001b[1;32m    163\u001b[0m     )\n\u001b[1;32m    165\u001b[0m     final_outputs: Dict[\u001b[38;5;28mstr\u001b[39m, Any] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprep_outputs(\n\u001b[1;32m    166\u001b[0m         inputs, outputs, return_only_outputs\n\u001b[1;32m    167\u001b[0m     )\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/llm.py:126\u001b[0m, in \u001b[0;36mLLMChain._call\u001b[0;34m(self, inputs, run_manager)\u001b[0m\n\u001b[1;32m    121\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_call\u001b[39m(\n\u001b[1;32m    122\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    123\u001b[0m     inputs: Dict[\u001b[38;5;28mstr\u001b[39m, Any],\n\u001b[1;32m    124\u001b[0m     run_manager: Optional[CallbackManagerForChainRun] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    125\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Dict[\u001b[38;5;28mstr\u001b[39m, \u001b[38;5;28mstr\u001b[39m]:\n\u001b[0;32m--> 126\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43minputs\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_manager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    127\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcreate_outputs(response)[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain/chains/llm.py:138\u001b[0m, in \u001b[0;36mLLMChain.generate\u001b[0;34m(self, input_list, run_manager)\u001b[0m\n\u001b[1;32m    136\u001b[0m callbacks \u001b[38;5;241m=\u001b[39m run_manager\u001b[38;5;241m.\u001b[39mget_child() \u001b[38;5;28;01mif\u001b[39;00m run_manager \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    137\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mllm, BaseLanguageModel):\n\u001b[0;32m--> 138\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mllm\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate_prompt\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    139\u001b[0m \u001b[43m        \u001b[49m\u001b[43mprompts\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    140\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    141\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    142\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mllm_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    143\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    144\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    145\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mllm\u001b[38;5;241m.\u001b[39mbind(stop\u001b[38;5;241m=\u001b[39mstop, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mllm_kwargs)\u001b[38;5;241m.\u001b[39mbatch(\n\u001b[1;32m    146\u001b[0m         cast(List, prompts), {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcallbacks\u001b[39m\u001b[38;5;124m\"\u001b[39m: callbacks}\n\u001b[1;32m    147\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:786\u001b[0m, in \u001b[0;36mBaseChatModel.generate_prompt\u001b[0;34m(self, prompts, stop, callbacks, **kwargs)\u001b[0m\n\u001b[1;32m    778\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgenerate_prompt\u001b[39m(\n\u001b[1;32m    779\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    780\u001b[0m     prompts: \u001b[38;5;28mlist\u001b[39m[PromptValue],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    783\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[1;32m    784\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m LLMResult:\n\u001b[1;32m    785\u001b[0m     prompt_messages \u001b[38;5;241m=\u001b[39m [p\u001b[38;5;241m.\u001b[39mto_messages() \u001b[38;5;28;01mfor\u001b[39;00m p \u001b[38;5;129;01min\u001b[39;00m prompts]\n\u001b[0;32m--> 786\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprompt_messages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallbacks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcallbacks\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:643\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[0;34m(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)\u001b[0m\n\u001b[1;32m    641\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m run_managers:\n\u001b[1;32m    642\u001b[0m             run_managers[i]\u001b[38;5;241m.\u001b[39mon_llm_error(e, response\u001b[38;5;241m=\u001b[39mLLMResult(generations\u001b[38;5;241m=\u001b[39m[]))\n\u001b[0;32m--> 643\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    644\u001b[0m flattened_outputs \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m    645\u001b[0m     LLMResult(generations\u001b[38;5;241m=\u001b[39m[res\u001b[38;5;241m.\u001b[39mgenerations], llm_output\u001b[38;5;241m=\u001b[39mres\u001b[38;5;241m.\u001b[39mllm_output)  \u001b[38;5;66;03m# type: ignore[list-item]\u001b[39;00m\n\u001b[1;32m    646\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m results\n\u001b[1;32m    647\u001b[0m ]\n\u001b[1;32m    648\u001b[0m llm_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_combine_llm_outputs([res\u001b[38;5;241m.\u001b[39mllm_output \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m results])\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:633\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[0;34m(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)\u001b[0m\n\u001b[1;32m    630\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(messages):\n\u001b[1;32m    631\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    632\u001b[0m         results\u001b[38;5;241m.\u001b[39mappend(\n\u001b[0;32m--> 633\u001b[0m             \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_generate_with_cache\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    634\u001b[0m \u001b[43m                \u001b[49m\u001b[43mm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    635\u001b[0m \u001b[43m                \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    636\u001b[0m \u001b[43m                \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_managers\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mrun_managers\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    637\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    638\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    639\u001b[0m         )\n\u001b[1;32m    640\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    641\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m run_managers:\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_core/language_models/chat_models.py:851\u001b[0m, in \u001b[0;36mBaseChatModel._generate_with_cache\u001b[0;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    849\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    850\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m inspect\u001b[38;5;241m.\u001b[39msignature(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate)\u001b[38;5;241m.\u001b[39mparameters\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_manager\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m--> 851\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_generate\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    852\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmessages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrun_manager\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrun_manager\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m    853\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    854\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    855\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate(messages, stop\u001b[38;5;241m=\u001b[39mstop, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/langchain_openai/chat_models/base.py:707\u001b[0m, in \u001b[0;36mBaseChatOpenAI._generate\u001b[0;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[1;32m    705\u001b[0m     generation_info \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mheaders\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;28mdict\u001b[39m(raw_response\u001b[38;5;241m.\u001b[39mheaders)}\n\u001b[1;32m    706\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 707\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mpayload\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    708\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_create_chat_result(response, generation_info)\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/openai/_utils/_utils.py:274\u001b[0m, in \u001b[0;36mrequired_args.<locals>.inner.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    272\u001b[0m             msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMissing required argument: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mquote(missing[\u001b[38;5;241m0\u001b[39m])\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    273\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(msg)\n\u001b[0;32m--> 274\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/openai/resources/chat/completions.py:815\u001b[0m, in \u001b[0;36mCompletions.create\u001b[0;34m(self, messages, model, audio, frequency_penalty, function_call, functions, logit_bias, logprobs, max_completion_tokens, max_tokens, metadata, modalities, n, parallel_tool_calls, presence_penalty, response_format, seed, service_tier, stop, store, stream, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)\u001b[0m\n\u001b[1;32m    775\u001b[0m \u001b[38;5;129m@required_args\u001b[39m([\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessages\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel\u001b[39m\u001b[38;5;124m\"\u001b[39m], [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessages\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m    776\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcreate\u001b[39m(\n\u001b[1;32m    777\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    812\u001b[0m     timeout: \u001b[38;5;28mfloat\u001b[39m \u001b[38;5;241m|\u001b[39m httpx\u001b[38;5;241m.\u001b[39mTimeout \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m|\u001b[39m NotGiven \u001b[38;5;241m=\u001b[39m NOT_GIVEN,\n\u001b[1;32m    813\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ChatCompletion \u001b[38;5;241m|\u001b[39m Stream[ChatCompletionChunk]:\n\u001b[1;32m    814\u001b[0m     validate_response_format(response_format)\n\u001b[0;32m--> 815\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_post\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    816\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/chat/completions\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    817\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmaybe_transform\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    818\u001b[0m \u001b[43m            \u001b[49m\u001b[43m{\u001b[49m\n\u001b[1;32m    819\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmessages\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmessages\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    820\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmodel\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    821\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43maudio\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43maudio\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    822\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfrequency_penalty\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfrequency_penalty\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    823\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfunction_call\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunction_call\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    824\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfunctions\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunctions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    825\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mlogit_bias\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mlogit_bias\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    826\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mlogprobs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mlogprobs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    827\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmax_completion_tokens\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmax_completion_tokens\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    828\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmax_tokens\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmax_tokens\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    829\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmetadata\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    830\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmodalities\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodalities\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    831\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mn\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    832\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mparallel_tool_calls\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mparallel_tool_calls\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    833\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpresence_penalty\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mpresence_penalty\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    834\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mresponse_format\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mresponse_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    835\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mseed\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mseed\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    836\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mservice_tier\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mservice_tier\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    837\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstop\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstop\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    838\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstore\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstore\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    839\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstream\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    840\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstream_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    841\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtemperature\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemperature\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    842\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtool_choice\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtool_choice\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    843\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtools\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtools\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    844\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtop_logprobs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtop_logprobs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    845\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtop_p\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtop_p\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    846\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43muser\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43muser\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    847\u001b[0m \u001b[43m            \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    848\u001b[0m \u001b[43m            \u001b[49m\u001b[43mcompletion_create_params\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mCompletionCreateParams\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    849\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    850\u001b[0m \u001b[43m        \u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmake_request_options\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    851\u001b[0m \u001b[43m            \u001b[49m\u001b[43mextra_headers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextra_headers\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mextra_query\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextra_query\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mextra_body\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mextra_body\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    852\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    853\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcast_to\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mChatCompletion\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    854\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    855\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstream_cls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mStream\u001b[49m\u001b[43m[\u001b[49m\u001b[43mChatCompletionChunk\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    856\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/openai/_base_client.py:1277\u001b[0m, in \u001b[0;36mSyncAPIClient.post\u001b[0;34m(self, path, cast_to, body, options, files, stream, stream_cls)\u001b[0m\n\u001b[1;32m   1263\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpost\u001b[39m(\n\u001b[1;32m   1264\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1265\u001b[0m     path: \u001b[38;5;28mstr\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1272\u001b[0m     stream_cls: \u001b[38;5;28mtype\u001b[39m[_StreamT] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m   1273\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ResponseT \u001b[38;5;241m|\u001b[39m _StreamT:\n\u001b[1;32m   1274\u001b[0m     opts \u001b[38;5;241m=\u001b[39m FinalRequestOptions\u001b[38;5;241m.\u001b[39mconstruct(\n\u001b[1;32m   1275\u001b[0m         method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpost\u001b[39m\u001b[38;5;124m\"\u001b[39m, url\u001b[38;5;241m=\u001b[39mpath, json_data\u001b[38;5;241m=\u001b[39mbody, files\u001b[38;5;241m=\u001b[39mto_httpx_files(files), \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39moptions\n\u001b[1;32m   1276\u001b[0m     )\n\u001b[0;32m-> 1277\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cast(ResponseT, \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcast_to\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mopts\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream_cls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream_cls\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/openai/_base_client.py:954\u001b[0m, in \u001b[0;36mSyncAPIClient.request\u001b[0;34m(self, cast_to, options, remaining_retries, stream, stream_cls)\u001b[0m\n\u001b[1;32m    951\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    952\u001b[0m     retries_taken \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m--> 954\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    955\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcast_to\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcast_to\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    956\u001b[0m \u001b[43m    \u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    957\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    958\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstream_cls\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream_cls\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    959\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries_taken\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries_taken\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    960\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/dl/lib/python3.10/site-packages/openai/_base_client.py:1058\u001b[0m, in \u001b[0;36mSyncAPIClient._request\u001b[0;34m(self, cast_to, options, retries_taken, stream, stream_cls)\u001b[0m\n\u001b[1;32m   1055\u001b[0m         err\u001b[38;5;241m.\u001b[39mresponse\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m   1057\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRe-raising status error\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1058\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_status_error_from_response(err\u001b[38;5;241m.\u001b[39mresponse) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1060\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_process_response(\n\u001b[1;32m   1061\u001b[0m     cast_to\u001b[38;5;241m=\u001b[39mcast_to,\n\u001b[1;32m   1062\u001b[0m     options\u001b[38;5;241m=\u001b[39moptions,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1066\u001b[0m     retries_taken\u001b[38;5;241m=\u001b[39mretries_taken,\n\u001b[1;32m   1067\u001b[0m )\n",
      "\u001b[0;31mNotFoundError\u001b[0m: Error code: 404 - {'error': {'message': 'The model `gpt-4o` does not exist or you do not have access to it.', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_found'}}"
     ]
    }
   ],
   "source": [
    "result = qa.run(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "answer  = parser.invoke(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnimalList(animals=[Animal(species='mouse', strain='SOD1G93A', gender='male', n_treatment=15, n_control=17), Animal(species='mouse', strain='SOD1G93A; LanCL1 cKI', gender='male', n_treatment=15, n_control=15), Animal(species='mouse', strain='SOD1G93A', gender='female', n_treatment=21, n_control=18), Animal(species='mouse', strain='SOD1G93A; LanCL1 cKI', gender='female', n_treatment=16, n_control=16)])"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "class AnimalDetails(BaseModel):\n",
    "    treatment: str = Field(description=\"What type of treatment or intervention are used?\")\n",
    "    way_of_administration: str = Field(description=\"What way of administation are used?\")\n",
    "    age_at_start: int = Field(description=\"Age of the start of treamtment\")\n",
    "    duration_unit: str = Field(description=\"In which units age of the start was Month/Week/Day and e.t.c\")\n",
    "    dosage: str = Field(description=\"Dosage of administration\")\n",
    "class AnimalDetailsList(BaseModel):\n",
    "    animal_details: List[AnimalDetails]\n",
    "parser2 = PydanticOutputParser(pydantic_object=AnimalDetailsList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "animal_descriptions = [\n",
    "    f\"{animal.gender} {animal.species} {animal.group} {animal.strain}\" \n",
    "    for animal in answer.animals\n",
    "]\n",
    "all_animals_description = \", \".join(animal_descriptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'male mouse G93A SOD1G93A, male mouse G93A; LanCL1 cKI SOD1G93A, female mouse G93A SOD1G93A, female mouse G93A; LanCL1 cKI SOD1G93A'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_animals_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "intro = f\"\"\"\n",
    "You are an assistant. Use the following information to answer the question very shortly\n",
    "Describe what intervention is used for each groups of animals (separated by ,): {all_animals_description}\n",
    "Give an answer in proper JSON format using double quotes around keys and values. \n",
    "\"\"\"\n",
    "prompt_template2 = intro+\"\"\" \n",
    "For example: \n",
    "{{\n",
    "  \"animal_details\": [\n",
    "    {{\n",
    "      \"treatment\": \"treatment1\", # short name of the treatment\n",
    "      \"way_of_administration\": \"way_of_administration1\",# Food, Intravenous, Water, Intraperitoneal, Genomic and e.t.c\n",
    "      \"age_at_start\": 2,#write only value for example 2 (second month of the life)\n",
    "      \"duration_unit\": \"Months\", # Year, Month, Week, Day and e.t.c if age_at_start equal to 0 then write here Days\n",
    "      \"dosage\": \"dosage1\"#only doage values\n",
    "    }},\n",
    "    {{\n",
    "      \"treatment\": \"treatment2\",#short name of the treatment\n",
    "      \"way_of_administration\": \"way_of_administration2\",# Food, Intravenous, Water, Intraperitoneal,Genomic and e.t.c\n",
    "      \"age_at_start\": 0, #write only value for example 2 (second month of the life)\n",
    "      \"duration_unit\": \"Days\", # Year, Month, Week, Day and e.t.c if age_at_start equal to 0 then write here Days\n",
    "      \"dosage\": \"dosage2\"#only doage values\n",
    "    }}\n",
    "  ]\n",
    "}}\n",
    "Context: {context}\n",
    "Question: {question}\n",
    "Answer:\n",
    "\"\"\"\n",
    "\n",
    "# Create a PromptTemplate instance, note the use of both 'context' and 'query'\n",
    "prompt2 = PromptTemplate(template=prompt_template, input_variables=[\"query\",\"context\"],\n",
    "                        partial_variables={\"format_instructions\": parser2.model_json_schema()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "animal_details=[AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A'), AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A'), AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A'), AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A')]\n"
     ]
    }
   ],
   "source": [
    "query2 = f\"\"\"\n",
    "What treatment or intervention or manipulation are used for each groups of animals: {all_animals_description}?\n",
    "\"\"\"\n",
    "\n",
    "prompt2 = PromptTemplate(\n",
    "    template=prompt_template2,\n",
    "    input_variables=[\"query\", \"context\"],\n",
    "    partial_variables={\"format_instructions\": parser2.model_json_schema()},\n",
    ")\n",
    "\n",
    "qa2 = RetrievalQA.from_chain_type(\n",
    "    llm=llm,\n",
    "    chain_type=\"stuff\",\n",
    "    retriever=compression_retriever,\n",
    "    chain_type_kwargs={\"prompt\": prompt2, \"document_variable_name\": \"context\"},\n",
    ")\n",
    "result2 = qa2.run(query=query2)\n",
    "answer2 = parser2.invoke(result2)\n",
    "print(answer2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnimalDetailsList(animal_details=[AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A'), AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A'), AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A'), AnimalDetails(treatment='Genomic', way_of_administration='Genomic', age_at_start=0, duration_unit='Days', dosage='N/A')])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answer2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "class AnimalResults(BaseModel):\n",
    "    median_treatment: Optional[float] = Field(description=\"Median treatment duration in units\")\n",
    "    max_treatment: Optional[float] = Field(description=\"Max treatment duration in units\")\n",
    "    treatment_units: str = Field(description=\"In what units measured lifespan\")\n",
    "    p_value: Optional[str] = Field(description=\"p-value for statistical analysis\")\n",
    "class AnimalResultsList(BaseModel):\n",
    "    animal_results: List[AnimalResults]\n",
    "parser3 = PydanticOutputParser(pydantic_object=AnimalResultsList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "intro = f\"\"\"\n",
    "Write lifespan results for each group of animals: {all_animals_description}\n",
    "Give an answer in proper JSON format using double quotes around keys and values. \n",
    "\"\"\"\n",
    "prompt_template3 = intro+\"\"\"\n",
    "For example: \n",
    "{{\n",
    "  \"animal_results\": [\n",
    "    {{\n",
    "      \"median_treatment\": 10.5, # median treatment lifespan of the group (only value)\n",
    "      \"max_treatment\": 15.3,# max treatment lifespan of the group (only value)\n",
    "      \"treatment_units\":\"treatment_units1\" # In what units measured lifespan Month, Age, Week\n",
    "      \"p_value\":0.01 #p-value of statistical test if exist (only value)\n",
    "    }}\n",
    "  ]\n",
    "}}\n",
    "Context: {context}\n",
    "Question: {question}\n",
    "Answer:\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "animal_results=[AnimalResults(median_treatment=141.0, max_treatment=None, treatment_units='days', p_value=None), AnimalResults(median_treatment=163.0, max_treatment=None, treatment_units='days', p_value='0.01'), AnimalResults(median_treatment=146.0, max_treatment=None, treatment_units='days', p_value=None), AnimalResults(median_treatment=168.0, max_treatment=None, treatment_units='days', p_value='0.01')]\n"
     ]
    }
   ],
   "source": [
    "query3 = f\"\"\"\n",
    "Lifespan or survival curve/results for {all_animals_description}\n",
    "\"\"\"\n",
    "\n",
    "prompt3 = PromptTemplate(\n",
    "    template=prompt_template3,\n",
    "    input_variables=[\"query\", \"context\"],\n",
    "    partial_variables={\"format_instructions\": parser3.model_json_schema()},\n",
    ")\n",
    "\n",
    "qa3 = RetrievalQA.from_chain_type(\n",
    "    llm=llm,\n",
    "    chain_type=\"stuff\",\n",
    "    retriever=compression_retriever,\n",
    "    chain_type_kwargs={\"prompt\": prompt3, \"document_variable_name\": \"context\"},\n",
    ")\n",
    "result3 = qa3.run(query3)\n",
    "answer3 = parser3.invoke(result3)\n",
    "print(answer3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_df = pd.DataFrame(index=[x for x in range(len(answer.animals))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_df['file'] = fule_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, anim in enumerate(answer.animals):\n",
    "\n",
    "    temp_df.loc[i, \"species\"] = anim.species\n",
    "    temp_df.loc[i, \"strain\"] = anim.strain\n",
    "    temp_df.loc[i, \"gender\"] = anim.gender\n",
    "    temp_df.loc[i, \"n_treatment\"] = anim.n_treatment\n",
    "    temp_df.loc[i, \"n_control\"] = anim.n_control\n",
    "\n",
    "for i, anim in enumerate(answer2.animal_details):\n",
    "    temp_df.loc[i, \"treatment\"] = anim.treatment\n",
    "    temp_df.loc[i, \"way_of_administration\"] = anim.way_of_administration\n",
    "    temp_df.loc[i, \"age_at_start\"] = anim.age_at_start\n",
    "    temp_df.loc[i, \"duration_unit\"] = anim.duration_unit\n",
    "    temp_df.loc[i, \"dosage\"] = anim.dosage\n",
    "for i, anim in enumerate(answer3.animal_results):\n",
    "    temp_df.loc[i, \"median_treatment\"] = anim.median_treatment\n",
    "    temp_df.loc[i, \"max_treatment\"] = anim.max_treatment\n",
    "    temp_df.loc[i, \"p_value\"] = anim.p_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = pd.concat([result_df,temp_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnimalList(animals=[Animal(species='mouse', strain='SOD1G93A', gender='male', n_treatment=15, n_control=17), Animal(species='mouse', strain='SOD1G93A; LanCL1 cKI', gender='male', n_treatment=15, n_control=15), Animal(species='mouse', strain='SOD1G93A', gender='female', n_treatment=21, n_control=18), Animal(species='mouse', strain='SOD1G93A; LanCL1 cKI', gender='female', n_treatment=16, n_control=16)])"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "answer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = result_df[\n",
    "    (result_df[\"treatment\"].notna())\n",
    "    & (result_df[\"way_of_administration\"].notna())\n",
    "    & (result_df[\"age_at_start\"].notna())\n",
    "    & (result_df[\"duration_unit\"].notna())\n",
    "    & (result_df[\"p_value\"].notna())\n",
    "    & (result_df[\"treatment\"]!=\"Control\")\n",
    "    \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>species</th>\n",
       "      <th>strain</th>\n",
       "      <th>gender</th>\n",
       "      <th>n_treatment</th>\n",
       "      <th>n_control</th>\n",
       "      <th>treatment</th>\n",
       "      <th>way_of_administration</th>\n",
       "      <th>age_at_start</th>\n",
       "      <th>duration_unit</th>\n",
       "      <th>dosage</th>\n",
       "      <th>median_treatment</th>\n",
       "      <th>max_treatment</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>s41418-019-0422-6</td>\n",
       "      <td>mouse</td>\n",
       "      <td>SOD1G93A; LanCL1 cKI</td>\n",
       "      <td>male</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Days</td>\n",
       "      <td>N/A</td>\n",
       "      <td>163.0</td>\n",
       "      <td>None</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>s41418-019-0422-6</td>\n",
       "      <td>mouse</td>\n",
       "      <td>SOD1G93A; LanCL1 cKI</td>\n",
       "      <td>female</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Days</td>\n",
       "      <td>N/A</td>\n",
       "      <td>168.0</td>\n",
       "      <td>None</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                file species                strain  gender  n_treatment  \\\n",
       "1  s41418-019-0422-6   mouse  SOD1G93A; LanCL1 cKI    male         15.0   \n",
       "3  s41418-019-0422-6   mouse  SOD1G93A; LanCL1 cKI  female         16.0   \n",
       "\n",
       "   n_control treatment way_of_administration  age_at_start duration_unit  \\\n",
       "1       15.0   Genomic               Genomic           0.0          Days   \n",
       "3       16.0   Genomic               Genomic           0.0          Days   \n",
       "\n",
       "  dosage  median_treatment max_treatment p_value  \n",
       "1    N/A             163.0          None    0.01  \n",
       "3    N/A             168.0          None    0.01  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.to_csv(f'{fule_name}.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = pd.DataFrame()\n",
    "for file in os.listdir(\"results\"):\n",
    "    file_path = \"results\" + \"/\" + file\n",
    "    temp_df = pd.read_csv(file_path)\n",
    "    result_df = pd.concat([result_df,temp_df],ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = result_df.rename(columns={'treatment':'intervention'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.to_excel(\"lifespangpt_results_v1.xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(\"CollidaData_2023.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 415,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cohort_id</th>\n",
       "      <th>volunteer_name</th>\n",
       "      <th>volunteer_country</th>\n",
       "      <th>study_id</th>\n",
       "      <th>doi</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>title</th>\n",
       "      <th>journal</th>\n",
       "      <th>year</th>\n",
       "      <th>full_text_URL</th>\n",
       "      <th>...</th>\n",
       "      <th>max_treatment</th>\n",
       "      <th>n_treatment</th>\n",
       "      <th>median_control</th>\n",
       "      <th>max_control</th>\n",
       "      <th>n_control</th>\n",
       "      <th>p_value</th>\n",
       "      <th>comment</th>\n",
       "      <th>timestamp</th>\n",
       "      <th>trust</th>\n",
       "      <th>extraction_method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>53</td>\n",
       "      <td>Leon Peshkin</td>\n",
       "      <td>France</td>\n",
       "      <td>595</td>\n",
       "      <td>10.1371/journal.pone.0083988</td>\n",
       "      <td>24409289</td>\n",
       "      <td>Mice fed rapamycin have an increase in lifespa...</td>\n",
       "      <td>PloS one</td>\n",
       "      <td>2014</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...</td>\n",
       "      <td>...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>32.5</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-05 02:46:28</td>\n",
       "      <td>2</td>\n",
       "      <td>Collida+DrugAge+ReviewbyEdouard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>51</td>\n",
       "      <td>Leon Peshkin</td>\n",
       "      <td>France</td>\n",
       "      <td>595</td>\n",
       "      <td>10.1371/journal.pone.0083988</td>\n",
       "      <td>24409289</td>\n",
       "      <td>Mice fed rapamycin have an increase in lifespa...</td>\n",
       "      <td>PloS one</td>\n",
       "      <td>2014</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...</td>\n",
       "      <td>...</td>\n",
       "      <td>38.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>31.5</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-05 02:42:26</td>\n",
       "      <td>2</td>\n",
       "      <td>Collida+DrugAge+ReviewbyEdouard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>140</td>\n",
       "      <td>Nickolai Leschov</td>\n",
       "      <td>Russia</td>\n",
       "      <td>594</td>\n",
       "      <td>10.1159/000212659</td>\n",
       "      <td>6519438</td>\n",
       "      <td>Dietary vitamin C improves the survival of mice</td>\n",
       "      <td>Gerontology</td>\n",
       "      <td>1984</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...</td>\n",
       "      <td>...</td>\n",
       "      <td>993.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>859.0</td>\n",
       "      <td>965.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>It should be noted that the ascorbic acid grou...</td>\n",
       "      <td>2023-02-27 17:42:25</td>\n",
       "      <td>2</td>\n",
       "      <td>Collida+DrugAge+ReviewbyEdouard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>42</td>\n",
       "      <td>Edouard Debonneuil</td>\n",
       "      <td>France</td>\n",
       "      <td>593</td>\n",
       "      <td>10.1073/pnas.1717065115</td>\n",
       "      <td>29378959</td>\n",
       "      <td>Effects of rapamycin on growth hormone recepto...</td>\n",
       "      <td>Proceedings of the National Academy of Science...</td>\n",
       "      <td>2018</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...</td>\n",
       "      <td>...</td>\n",
       "      <td>1090.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>1275.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-04 17:21:52</td>\n",
       "      <td>2</td>\n",
       "      <td>Collida+DrugAge+ReviewbyEdouard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>50</td>\n",
       "      <td>Edouard Debonneuil</td>\n",
       "      <td>France</td>\n",
       "      <td>593</td>\n",
       "      <td>10.1073/pnas.1717065115</td>\n",
       "      <td>29378959</td>\n",
       "      <td>Effects of rapamycin on growth hormone recepto...</td>\n",
       "      <td>Proceedings of the National Academy of Science...</td>\n",
       "      <td>2018</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...</td>\n",
       "      <td>...</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1110.0</td>\n",
       "      <td>1275.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-02-04 18:00:54</td>\n",
       "      <td>2</td>\n",
       "      <td>Collida+DrugAge+ReviewbyEdouard</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   cohort_id      volunteer_name volunteer_country  study_id  \\\n",
       "0         53        Leon Peshkin            France       595   \n",
       "1         51        Leon Peshkin            France       595   \n",
       "2        140    Nickolai Leschov            Russia       594   \n",
       "3         42  Edouard Debonneuil            France       593   \n",
       "4         50  Edouard Debonneuil            France       593   \n",
       "\n",
       "                            doi  pubmed_id  \\\n",
       "0  10.1371/journal.pone.0083988   24409289   \n",
       "1  10.1371/journal.pone.0083988   24409289   \n",
       "2             10.1159/000212659    6519438   \n",
       "3       10.1073/pnas.1717065115   29378959   \n",
       "4       10.1073/pnas.1717065115   29378959   \n",
       "\n",
       "                                               title  \\\n",
       "0  Mice fed rapamycin have an increase in lifespa...   \n",
       "1  Mice fed rapamycin have an increase in lifespa...   \n",
       "2    Dietary vitamin C improves the survival of mice   \n",
       "3  Effects of rapamycin on growth hormone recepto...   \n",
       "4  Effects of rapamycin on growth hormone recepto...   \n",
       "\n",
       "                                             journal  year  \\\n",
       "0                                           PloS one  2014   \n",
       "1                                           PloS one  2014   \n",
       "2                                        Gerontology  1984   \n",
       "3  Proceedings of the National Academy of Science...  2018   \n",
       "4  Proceedings of the National Academy of Science...  2018   \n",
       "\n",
       "                                       full_text_URL  ... max_treatment  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...  ...          36.0   \n",
       "1  https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...  ...          38.0   \n",
       "2  https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...  ...         993.0   \n",
       "3  https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...  ...        1090.0   \n",
       "4  https://www.ncbi.nlm.nih.gov/pmc/articles/pmid...  ...        1150.0   \n",
       "\n",
       "  n_treatment median_control  max_control n_control p_value  \\\n",
       "0        45.0           27.0         32.5      45.0       0   \n",
       "1        45.0           29.0         31.5      45.0       0   \n",
       "2        16.0          859.0        965.0       8.0    0.01   \n",
       "3        27.0         1100.0       1275.0      20.0  0.0001   \n",
       "4        32.0         1110.0       1275.0      33.0   0.001   \n",
       "\n",
       "                                             comment           timestamp  \\\n",
       "0                                                NaN 2023-02-05 02:46:28   \n",
       "1                                                NaN 2023-02-05 02:42:26   \n",
       "2  It should be noted that the ascorbic acid grou... 2023-02-27 17:42:25   \n",
       "3                                                NaN 2023-02-04 17:21:52   \n",
       "4                                                NaN 2023-02-04 18:00:54   \n",
       "\n",
       "  trust                extraction_method  \n",
       "0     2  Collida+DrugAge+ReviewbyEdouard  \n",
       "1     2  Collida+DrugAge+ReviewbyEdouard  \n",
       "2     2  Collida+DrugAge+ReviewbyEdouard  \n",
       "3     2  Collida+DrugAge+ReviewbyEdouard  \n",
       "4     2  Collida+DrugAge+ReviewbyEdouard  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 415,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 416,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['10.1371/journal.pone.0083988', '10.1159/000212659',\n",
       "       '10.1073/pnas.1717065115', '10.1016/j.ebiom.2018.09.015',\n",
       "       '10.1016/j.cmet.2020.08.004', '10.1016/0531-5565(75)90012-1',\n",
       "       '10.1007/s10522-007-9100-z'], dtype=object)"
      ]
     },
     "execution_count": 416,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['extraction_method']=='Collida+DrugAge+ReviewbyEdouard']['doi'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['files'] = [x.split('/')[1] for x in df['doi']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = os.listdir('processed_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import jaccard_score,mean_absolute_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list = result_df['file'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 422,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'j.celrep.2013.07.030'"
      ]
     },
     "execution_count": 422,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_list[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_temp_df = result_df[result_df['file']==file_list[5]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 534,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_temp_df = df[df['files']==file_list[5]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['10.1073/pnas.1717065115'], dtype=object)"
      ]
     },
     "execution_count": 535,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_temp_df['doi'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminal_columns = ['species','strain','gender','n_treatment','n_control']\n",
    "animal_details_columns = ['intervention','way_of_administration','age_at_start','duration_unit','dosage']\n",
    "animal_results_columns = ['median_treatment','max_treatment','p_value']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_answer = result_temp_df[\"median_treatment\"].tolist()\n",
    "valid_answer = valid_temp_df['median_treatment'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[975.0, 1056.0]"
      ]
     },
     "execution_count": 538,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_answer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[955.0, 1055.0, 1055.0]"
      ]
     },
     "execution_count": 539,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_answer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 516,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_answer = [\"40\", \"40\",\"40\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 517,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_answer = [\"30.0\",\"30.0\",\"0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 519,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4444444444444444"
      ]
     },
     "execution_count": 519,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jaccard_score(valid_answer,result_answer,average=\"weighted\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['mice', 'mice', 'mice']"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_temp_df['species'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dl",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
